{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "46358dd6",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "# Proteogenomics database generation using py-pgatk\n",
    "\n",
    "\n",
    "This notebook presents commands in `py-pgatk` to:\n",
    "- download data from various sources including ENSEMBL for reference files, COSMIC & cBioPortal for cancer mutations, and gnomAD & ENSEMBL for population variants.\n",
    "- generate protein databases using the downloaded files or from custom files (mutations in VCF and transcripts in GTF)\n",
    "- explore the database content (number of proteins and overlaps in tryptic peptides between various protein categories)\n",
    "\n",
    "`py-pgatk` is implemented in `Python` and is available here: https://github.com/bigbio/py-pgatk. \n",
    "The commands are documented here: https://pgatk.readthedocs.io/en/latest/pypgatk.html.\n",
    "The tools are also builtin to a nextflow workflow: https://nf-co.re/pgdb.\n",
    "\n",
    "The easiset way to run this notebook is by installing conda and creating an environment for installing the required packages. \n",
    "\n",
    "\n",
    "# Part 1: Installation\n",
    "\n",
    "### Get notebook and data\n",
    "Updated version can be found here: https://github.com/husensofteng/ProteoWorks/blob/master/workshop_pgdb/pgdb.ipynb\n",
    "\n",
    "\n",
    "### Install Conda\n",
    "Download Conda for your operating system and follow the installation instructions from here: https://docs.conda.io/projects/conda/en/latest/user-guide/install/index.html#regular-installation\n",
    "\n",
    "Next, open a terminal window and run the following commands:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b69ff0c",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "source": [
    "Choose either `conda` (supports up to v.0.019) or `pip` that also has v0.020:\n",
    "### Create `pgdb` environment (only once\n",
    "`conda create -n pgdb jupyterlab matplotlib`\n",
    "\n",
    "\n",
    "### Activate `pgdb` environment (everytime with new terminal/login)\n",
    "`conda activate pgdb`\n",
    "\n",
    "### Browse to your working directory and start Jupyter from there using\n",
    "`jupyter-notebook`\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "62bf88ac",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pip in /Users/husen/opt/anaconda3/envs/pygatk/lib/python3.8/site-packages (22.1.2)\n",
      "Requirement already satisfied: wget in /Users/husen/opt/anaconda3/envs/pygatk/lib/python3.8/site-packages (3.2)\n"
     ]
    }
   ],
   "source": [
    "!pip install --upgrade pip\n",
    "\n",
    "!pip install wget"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "945d79ab",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Currently using version: 0.0.23dev_0<develop>\n",
      "Currently using version: 0.0.23dev_0<develop>\n"
     ]
    }
   ],
   "source": [
    "def get_version():\n",
    "    !conda list > package_versions.txt\n",
    "    try:\n",
    "        pypgatk_version = [x.strip() for x in open('package_versions.txt').readlines() if 'pypgatk' in x][-1].replace(' ','').replace('pypgatk', '').replace('pypi_0pypi', '')\n",
    "        print('Currently using version: {}'.format(pypgatk_version))\n",
    "        return pypgatk_version\n",
    "    except IndexError:\n",
    "        return None\n",
    "    \n",
    "\n",
    "try:\n",
    "    import pypgatk\n",
    "except ModuleNotFoundError:\n",
    "    !pip install pypgatk\n",
    "    try:\n",
    "        import pypgatk\n",
    "        !pip install pypgatk\n",
    "    except ModuleNotFoundError:\n",
    "        try: #get the latest source code from the repo\n",
    "            !git clone https://github.com/bigbio/py-pgatk.git\n",
    "            !pip install .\n",
    "            import pypgatk\n",
    "        except ModuleNotFoundError:\n",
    "            print('Failed to installed pypgatk, exiting')\n",
    "\n",
    "\n",
    "#Get version of the currently installed pypgatk to obtain the right configuration files\n",
    "if not get_version():\n",
    "    !pip install pypgatk\n",
    "    if not get_version():\n",
    "        !git clone https://github.com/bigbio/py-pgatk.git\n",
    "        !pip install .\n",
    "pypgatk_version = get_version()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "249576c2",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "### Download configuration files and related tools\n",
    "\n",
    "#### Import required libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "56f3b79d",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "import sys\n",
    "from collections import Counter\n",
    "\n",
    "try:\n",
    "    from Bio import SeqIO\n",
    "except ModuleNotFoundError:\n",
    "    !conda install -c conda-forge biopython\n",
    "    from Bio import SeqIO\n",
    "    \n",
    "try:\n",
    "    import matplotlib.pyplot as plt\n",
    "except ModuleNotFoundError:\n",
    "    !conda install matplotlib\n",
    "    import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40634fdd",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "#### Get Configuaration Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "04b15661",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the files are available in config/ in the current directory:\n",
      "  /Users/husen/Documents/Karolinska Institutet/GRP_LehtioLab_bioinformatics_brains - General/ProteoWorks/workshop_pgdb\n"
     ]
    }
   ],
   "source": [
    "if not os.path.isdir('config'):\n",
    "    os.system('wget https://github.com/bigbio/py-pgatk/archive/refs/tags/v{}.zip'.format(pypgatk_version))\n",
    "    os.system('unzip v{}.zip'.format(pypgatk_version))\n",
    "    shutil.copytree('py-pgatk-{}/pypgatk/config'.format(pypgatk_version), 'config')\n",
    "if os.path.isdir('config'):\n",
    "    print('the files are available in config/ in the current directory:\\n ', os.getcwd())\n",
    "    try:\n",
    "        os.rmdir('py-pgatk-*')\n",
    "    except FileNotFoundError:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "45137cf8",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ensembl_config.yaml',\n",
       " 'ensembl_downloader_config.yaml',\n",
       " 'cbioportal_config.yaml',\n",
       " 'openms_analysis.yaml',\n",
       " '__init__.py',\n",
       " 'assemblies_conf.json',\n",
       " 'cosmic_config.yaml',\n",
       " 'protein_decoy.yaml']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.listdir('config')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6246aea8",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "### Install gffread v.0.11.7 \n",
    "This tool is needed to convert GTF (gene-transript annotation file) to FASTA (DNA sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d8610ecc",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gffread is ready to be used\n"
     ]
    }
   ],
   "source": [
    "if not os.path.isfile('gffread'):\n",
    "    if 'linux' in sys.platform:\n",
    "        os.system('wget https://github.com/gpertea/gffread/releases/download/v0.12.7/gffread-0.12.7.Linux_x86_64.tar.gz')\n",
    "        os.system('tar -xvzf gffread-0.12.7.Linux_x86_64.tar.gz')\n",
    "        shutil.move('gffread-0.12.7.Linux_x86_64/gffread', 'gffread')\n",
    "    elif 'darwin' in sys.platform:\n",
    "        os.system('wget https://github.com/gpertea/gffread/releases/download/v0.12.7/gffread-0.12.7.OSX_x86_64.tar.gz')\n",
    "        os.system('tar -xvzf gffread-0.12.7.OSX_x86_64.tar.gz')\n",
    "        shutil.move('gffread-0.12.7.OSX_x86_64/gffread', 'gffread')\n",
    "    else:\n",
    "        print('SORRY: only Linux and MAC are supported, maybe you need to try manually installing gffread')\n",
    "\n",
    "if os.path.isfile('gffread'):\n",
    "    print('gffread is ready to be used')\n",
    "    try:\n",
    "        os.rmdir('gffread-*')\n",
    "    except FileNotFoundError:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da109a48",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "#### Before begining, let's test if pypgatk is working, you should see a help message that lists the existing commands"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2fff0b92",
   "metadata": {
    "scrolled": false,
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n",
      "Usage: pypgatk_cli.py [OPTIONS] COMMAND [ARGS]...\n",
      "\n",
      "  This is the main tool that give access to all commands and options\n",
      "  provided by the pypgatk\n",
      "\n",
      "Options:\n",
      "  --version   Show the version and exit.\n",
      "  -h, --help  Show this message and exit.\n",
      "\n",
      "Commands:\n",
      "  cbioportal-downloader    Command to download the the cbioportal studies\n",
      "  cbioportal-to-proteindb  Command to translate cbioportal mutation data into\n",
      "                           proteindb\n",
      "  cosmic-downloader        Command to download the cosmic mutation database\n",
      "  cosmic-to-proteindb      Command to translate Cosmic mutation data into\n",
      "                           proteindb\n",
      "  dnaseq-to-proteindb      Generate peptides based on DNA sequences\n",
      "  ensembl-check            Command to check ensembl database for stop codons,\n",
      "                           gaps\n",
      "  ensembl-downloader       Command to download the ensembl information\n",
      "  generate-decoy           Create decoy protein sequences using multiple\n",
      "                           methods DecoyPYrat, Reverse/Shuffled Proteins.\n",
      "  generate-deeplc          Generate input for deepLC tool from idXML,mzTab or\n",
      "                           consensusXML\n",
      "  msrescore-configuration  Command to generate the msrescore configuration\n",
      "                           file from idXML\n",
      "  peptide-class-fdr        Command to compute the Peptide class FDR\n",
      "  threeframe-translation   Command to perform 3'frame translation\n",
      "  vcf-to-proteindb         Generate peptides based on DNA variants VCF files\n"
     ]
    }
   ],
   "source": [
    "!pypgatk_cli.py -h"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7064d93c",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "# Part 2: DNA to Protein\n",
    "\n",
    "In this part we will use the `dnaseq-to-proteindb` command to translate DNA sequences in fasta format to protein. We will also show how to convert a GTF obtained from a reference assembly or from a RNA-seq to DNA sequences and then protein sequence.\n",
    "\n",
    "## Download relevant files\n",
    "\n",
    "First we need to specify what files to download from the reference assembly. By default, `ensembl-downloader` downloads all files. The flags below can be used to ignore some files, e.g --skip_dna will skip downloading the reference DNA sequence, etc.\n",
    "\n",
    "* Note: versions prior to v0.0.20 don't have the feature to download the DNA reference assembly. IF you are using an older version you need to download the DNA fasta manually, uncomment the commands given below to perform this step."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1fcb9191",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "### Check the flags below to know what flags to use."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "15232464",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n",
      "Usage: pypgatk_cli.py ensembl-downloader [OPTIONS]\n",
      "\n",
      "  This tool enables to download from enseml ftp the FASTA and GTF files\n",
      "\n",
      "Options:\n",
      "  -c, --config_file TEXT          Configuration file for the ensembl data\n",
      "                                  downloader pipeline\n",
      "  -o, --output_directory TEXT     Output directory for the peptide databases\n",
      "  -fp, --folder_prefix_release TEXT\n",
      "                                  Output folder prefix to download the data\n",
      "  -t, --taxonomy TEXT             Taxonomy identifiers (comma separated list\n",
      "                                  can be given) that will be use to download\n",
      "                                  the data from Ensembl\n",
      "  -l, --list_taxonomies           Print the list of all the taxonomies in\n",
      "                                  ENSEMBL (https://www.ensembl.org)\n",
      "  -sg, --skip_gtf                 Skip the GTF file during the download\n",
      "  -sp, --skip_protein             Skip the protein fasta file during download\n",
      "  -sc, --skip_cds                 Skip the CDS file download\n",
      "  -sdn, --skip_cdna               Skip the cDNA file download\n",
      "  -sn, --skip_ncrna               Skip the ncRNA file download\n",
      "  -sd, --skip_dna                 Skip the DNA (reference genome assembly)\n",
      "                                  file download\n",
      "  -sv, --skip_vcf                 Skip the VCF variant file\n",
      "  -en, --ensembl_name TEXT        Ensembl name code to download, it can be use\n",
      "                                  instead of taxonomy (e.g. homo_sapiens)\n",
      "  --grch37                        Download a previous version GRCh37 of\n",
      "                                  ensembl genomes\n",
      "  --url_file TEXT                 Add the url to a downloaded file\n",
      "  -h, --help                      Show this message and exit.\n"
     ]
    }
   ],
   "source": [
    "!pypgatk_cli.py ensembl-downloader -h"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfe4e0d8",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "\n",
    "#### Get taxonomy ID for the species of interest\n",
    "However, as shown above we need to specify a taxonomy number in order to get files from a reference assembly for our species of interest. \n",
    "\n",
    "Can you find the taxonomy ID for homo sapiens using the following command?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57c9a0ea",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [],
   "source": [
    "!pypgatk_cli.py ensembl-downloader -l -sv -sg -sp -sc -sd -sn "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f6f0d19",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "Let's download reference files except VCF (variants) for `Meleagris_gallopavo.Turkey` (taxonomy id: 9103) into a folder named ensembl_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a3e52896",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n",
      "Downloading --  ftp://ftp.ensembl.org/pub/release-106/fasta/meleagris_gallopavo/pep/Meleagris_gallopavo.Turkey_5.1.pep.all.fa.gz\n",
      "DEBUG:pypgatk.toolbox.general:Files downloaded -- ensembl_files/Meleagris_gallopavo.Turkey_5.1.pep.all.fa,ensembl_files/Meleagris_gallopavo.Turkey_5.1.106.gtf,ensembl_files/Meleagris_gallopavo.Turkey_5.1.cds.all.fa,ensembl_files/Meleagris_gallopavo.Turkey_5.1.cdna.all.fa,ensembl_files/Meleagris_gallopavo.Turkey_5.1.ncrna.fa,ensembl_files/Meleagris_gallopavo.Turkey_5.1.dna_sm.toplevel.fa\n",
      "INFO:Main Pipeline Ensembl Downloader:Pipeline Finish !!!\n"
     ]
    }
   ],
   "source": [
    "!pypgatk_cli.py ensembl-downloader -t 9103 -o ensembl_files --skip_vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ae49c10",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "source": [
    "Check the contents of the ensembl_files on your computer system, you should see the following files downloaded:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bd4bf472",
   "metadata": {
    "scrolled": true,
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Homo_sapiens.GRCh38.106.db',\n",
       " 'Meleagris_gallopavo.Turkey_5.1.106.gtf',\n",
       " 'Meleagris_gallopavo.Turkey_5.1.dna_sm.toplevel.fa',\n",
       " 'Meleagris_gallopavo.Turkey_5.1.pep.all.fa',\n",
       " 'Homo_sapiens.GRCh38.106.gtf',\n",
       " 'Homo_sapiens.GRCh38.cds.all.fa',\n",
       " 'Meleagris_gallopavo.Turkey_5.1.ncrna.fa',\n",
       " 'Meleagris_gallopavo.Turkey_5.1.cds.all.fa',\n",
       " 'Meleagris_gallopavo.Turkey_5.1.cdna.all.fa',\n",
       " 'Meleagris_gallopavo.Turkey_5.1.dna_sm.toplevel.fa.fai',\n",
       " 'Meleagris_gallopavo.Turkey_5.1.106.fa']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.listdir('ensembl_files')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4aa27dc7",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "#### Convert GTF (can be reference or custom from RNA-seq) to DNA sequences\n",
    "Although the cdna.fa and ncrna.fa files can be used as reference DNA, but we wull use  GTF since it provides more flexibility to select desired transcript types. Also, when running your own RNA-seq experiments you usually get a GTF file as a final output therefore it is better to learn how to translate transcripts starting from a GTF file.\n",
    "\n",
    "However, since the `dnaseq-to-proteindb` tool only accepts FASTA files as input, we first need to convert the GTF into FASTA using gffread that is a separate tool.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2254ac9b",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FASTA index file ensembl_files/Meleagris_gallopavo.Turkey_5.1.dna_sm.toplevel.fa.fai created.\r\n"
     ]
    }
   ],
   "source": [
    "!./gffread -F -w ensembl_files/Meleagris_gallopavo.Turkey_5.1.106.fa -g ensembl_files/Meleagris_gallopavo.Turkey_5.1.dna_sm.toplevel.fa ensembl_files/Meleagris_gallopavo.Turkey_5.1.106.gtf\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f6ceddc",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "Now that we have the DNA fasta file ready to be used, let's take a look at the parameters that can be used:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0e23d677",
   "metadata": {
    "scrolled": true,
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n",
      "Usage: pypgatk_cli.py dnaseq-to-proteindb [OPTIONS]\n",
      "\n",
      "Options:\n",
      "  -c, --config_file TEXT          Configuration to perform conversion between\n",
      "                                  ENSEMBL Files\n",
      "  --input_fasta TEXT              Path to sequences fasta\n",
      "  --translation_table INTEGER     Translation Table (default 1)\n",
      "  --num_orfs INTEGER              Number of ORFs (default 3)\n",
      "  --num_orfs_complement INTEGER   Number of ORFs from the reverse side\n",
      "                                  (default 0)\n",
      "  --output_proteindb TEXT         Output file name, exits if already exists\n",
      "  -p, --var_prefix TEXT           String to add before the variant peptides\n",
      "  --skip_including_all_cds        By default any transcript that has a defined\n",
      "                                  CDS will be translated, this option disables\n",
      "                                  this features instead it only depends on the\n",
      "                                  biotypes\n",
      "  --include_biotypes TEXT         Include Biotypes\n",
      "  --exclude_biotypes TEXT         Exclude Biotypes\n",
      "  --biotype_str TEXT              String used to identify gene/transcript\n",
      "                                  biotype in the gtf file.\n",
      "  --transcript_description_sep TEXT\n",
      "                                  Separator used to separate features in the\n",
      "                                  fasta headers, usually either (space, / or\n",
      "                                  semicolon).\n",
      "  --expression_str TEXT           String to be used for extracting expression\n",
      "                                  value (TPM, FPKM, etc).\n",
      "  --expression_thresh FLOAT       Threshold used to filter transcripts based\n",
      "                                  on their expression values\n",
      "  -h, --help                      Show this message and exit.\n"
     ]
    }
   ],
   "source": [
    "!pypgatk_cli.py dnaseq-to-proteindb  -h"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3854d360",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "#### Generate a protein database from lncRNAs only\n",
    "*Use the include_biotypes flag to specify what kind of transcripts should be translated. See ensembl biotypes: http://www.ensembl.org/info/genome/genebuild/biotypes.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "71a84c4c",
   "metadata": {
    "scrolled": false,
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/husen/opt/anaconda3/envs/pygatk/bin/pypgatk_cli.py\", line 7, in <module>\n",
      "    exec(compile(f.read(), __file__, 'exec'))\n",
      "  File \"/Users/husen/git/py-pgatk/pypgatk/pypgatk_cli.py\", line 56, in <module>\n",
      "    main()\n",
      "  File \"/Users/husen/git/py-pgatk/pypgatk/pypgatk_cli.py\", line 52, in main\n",
      "    cli()\n",
      "  File \"/Users/husen/opt/anaconda3/envs/pygatk/lib/python3.8/site-packages/click/core.py\", line 764, in __call__\n",
      "    return self.main(*args, **kwargs)\n",
      "  File \"/Users/husen/opt/anaconda3/envs/pygatk/lib/python3.8/site-packages/click/core.py\", line 717, in main\n",
      "    rv = self.invoke(ctx)\n",
      "  File \"/Users/husen/opt/anaconda3/envs/pygatk/lib/python3.8/site-packages/click/core.py\", line 1137, in invoke\n",
      "    return _process_result(sub_ctx.command.invoke(sub_ctx))\n",
      "  File \"/Users/husen/opt/anaconda3/envs/pygatk/lib/python3.8/site-packages/click/core.py\", line 956, in invoke\n",
      "    return ctx.invoke(self.callback, **ctx.params)\n",
      "  File \"/Users/husen/opt/anaconda3/envs/pygatk/lib/python3.8/site-packages/click/core.py\", line 555, in invoke\n",
      "    return callback(*args, **kwargs)\n",
      "  File \"/Users/husen/opt/anaconda3/envs/pygatk/lib/python3.8/site-packages/click/decorators.py\", line 17, in new_func\n",
      "    return f(get_current_context(), *args, **kwargs)\n",
      "  File \"/Users/husen/git/py-pgatk/pypgatk/commands/dnaseq_to_proteindb.py\", line 84, in dnaseq_to_proteindb\n",
      "    ensembl_data_service = EnsemblDataService(config_data, pipeline_arguments)\n",
      "  File \"/Users/husen/git/py-pgatk/pypgatk/ensembl/ensembl.py\", line 49, in __init__\n",
      "    self._proteindb_output = self.get_default_parameters()[self.CONFIG_KEY_DATA][self.PROTEIN_DB_OUTPUT]\n",
      "KeyError: 'ensembl_translation'\n",
      "\n",
      "Generated: dict_items([])\n"
     ]
    }
   ],
   "source": [
    "!pypgatk_cli.py dnaseq-to-proteindb --input_fasta ensembl_files/Meleagris_gallopavo.Turkey_5.1.106.fa --output_proteindb lncRNAs.fa --skip_including_all_cds --include_biotypes lncRNA\n",
    "print(\"\\nGenerated:\", Counter([x.split('_')[0] for x in list(SeqIO.index('lncRNAs.fa', \"fasta\").keys())]).items())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6ce9b58",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "#### Generate protein database from other non-coding RNAs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e63a90af",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n",
      "/Users/husen/opt/anaconda3/envs/pygatk/lib/python3.8/site-packages/Bio/Seq.py:2606: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\"Partial codon, len(sequence) not a multiple of three. \"\n"
     ]
    }
   ],
   "source": [
    "!pypgatk_cli.py dnaseq-to-proteindb --config_file config/ensembl_config.yaml --input_fasta ensembl_files/Meleagris_gallopavo.Turkey_5.1.106.fa --output_proteindb ncRNAs.fa --skip_including_all_cds --var_prefix ncRNA_ --include_biotypes miRNA,miscRNA,piRNA,rRNA,siRNA,snRNA,snoRNA,tRNA,vaultRNA\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "159217e4",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Generated: dict_items([('ncRNA', 1500)])\n"
     ]
    }
   ],
   "source": [
    "print(\"\\nGenerated:\", Counter([x.split('_')[0] for x in list(SeqIO.index('ncRNAs.fa', \"fasta\").keys())]).items())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56d4bf8c",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "#### Generate protein database from alternative reading frames of coding sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "228362a6",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \r\n",
      " platform, measuring for memoryleaks will never fail\r\n"
     ]
    }
   ],
   "source": [
    "!pypgatk_cli.py dnaseq-to-proteindb --config_file config/ensembl_config.yaml --input_fasta ensembl_files/Meleagris_gallopavo.Turkey_5.1.106.fa --output_proteindb altORFs.fa --skip_including_all_cds --var_prefix altORF_ --include_biotypes altORFs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2c65cbaa",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Generated: dict_items([])\n"
     ]
    }
   ],
   "source": [
    "print(\"\\nGenerated:\", Counter([x.split('_')[0] for x in list(SeqIO.index('altORFs.fa', \"fasta\").keys())]).items())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88a8f5f6",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "#### Generate a protein database from pseudogenes? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "069f0fd3",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n",
      "/Users/husen/opt/anaconda3/envs/pygatk/lib/python3.8/site-packages/Bio/Seq.py:2606: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\"Partial codon, len(sequence) not a multiple of three. \"\n"
     ]
    }
   ],
   "source": [
    "#write your code here\n",
    "!pypgatk_cli.py dnaseq-to-proteindb --config_file config/ensembl_config.yaml --input_fasta ensembl_files/Meleagris_gallopavo.Turkey_5.1.106.fa --output_proteindb pesuedogenes.fa --skip_including_all_cds --var_prefix pesuedogene_ --include_biotypes pseudogene,processed_pseudogene\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "8c46a69d",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Generated: dict_items([('pesuedogene', 477)])\n"
     ]
    }
   ],
   "source": [
    "print(\"\\nGenerated:\", Counter([x.split('_')[0] for x in list(SeqIO.index('pesuedogenes.fa', \"fasta\").keys())]).items())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abe14dba",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "### Concatenate the generated protein databases with reference proteome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "590ff8aa",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_items([('lncRNA', 5109), ('pesuedogene', 477), ('ncRNA', 1500), ('Can', 28295)])"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "!cat lncRNAs.fa pesuedogenes.fa ncRNAs.fa altORFs.fa ensembl_files/Meleagris_gallopavo.Turkey_5.1.pep.all.fa > proteins.fa\n",
    "\n",
    "Counter([x.replace('ENS','Can_ENS').split('_')[0] for x in list(SeqIO.index('proteins.fa', \"fasta\").keys())]).items()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7aa09bc8",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "#### Perform six-frame translation on DNA seqyence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b204e3b4",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n",
      "/Users/husen/opt/anaconda3/envs/pygatk/lib/python3.8/site-packages/Bio/Seq.py:2606: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\"Partial codon, len(sequence) not a multiple of three. \"\n"
     ]
    }
   ],
   "source": [
    "!pypgatk_cli.py dnaseq-to-proteindb --config_file config/ensembl_config.yaml --skip_including_all_cds --input_fasta ensembl_files/Meleagris_gallopavo.Turkey_5.1.106.fa --output_proteindb sixframes.fa --var_prefix sixframe_  --num_orfs 3 --num_orfs_complement 3 --include_biotypes all\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e8df5f85",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Generated: dict_items([('sixframe', 184248)])\n"
     ]
    }
   ],
   "source": [
    "print(\"\\nGenerated:\", Counter([x.split('_')[0] for x in list(SeqIO.index('sixframes.fa', \"fasta\").keys())]).items())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "8c5a1f32",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [],
   "source": [
    "#function to digest proteins\n",
    "\n",
    "def trypsin_digestion(proseq_incl_stop, miss_cleavage=0):\n",
    "    \"digest peptides using the tryptic rule, allowing for miss cleavages\"\n",
    "    all_peptides = []\n",
    "    for protseq in proseq_incl_stop.split('*'):\n",
    "        if len(protseq)<=0:\n",
    "            continue\n",
    "        peptides = []\n",
    "        peptide = ''\n",
    "        \"remove the first K/R if in the begining of a reading frame\"\n",
    "        protseq_updated = protseq[0::]\n",
    "        if protseq[0]=='K' or protseq[0]=='R' and len(protseq)>1:\n",
    "            protseq_updated = protseq[1::]\n",
    "\n",
    "        for c, aa in enumerate(protseq_updated):\n",
    "            peptide += aa\n",
    "            next_aa = ''\n",
    "            try:\n",
    "                next_aa = protseq_updated[c + 1]\n",
    "            except IndexError:\n",
    "                pass\n",
    "\n",
    "            if aa in ['K', 'R'] and next_aa != 'P':  # for trypsin peptides\n",
    "                if len(peptide) > 0:\n",
    "                    peptides.append(peptide)\n",
    "                peptide = ''\n",
    "                continue\n",
    "\n",
    "        if len(peptide) > 0:\n",
    "            peptides.append(peptide)\n",
    "\n",
    "        peptides_with_miss_cleavage = []\n",
    "        for i in range(1, miss_cleavage + 1):\n",
    "            for j, pep in enumerate(peptides):\n",
    "                if j + i < len(peptides):\n",
    "                    peptide = ''.join([x for x in (peptides[j:j + i + 1])])\n",
    "                    peptides_with_miss_cleavage.append(peptide)\n",
    "\n",
    "        peptides.extend(peptides_with_miss_cleavage)\n",
    "        all_peptides.extend(peptides)\n",
    "\n",
    "    return all_peptides"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "300c2798",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "#### Investigate the overlap between tryptic peptides from various categories!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "d702f439",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of tryptic peptides per type:\n",
      " lncRNA = 170307\n",
      "pesuedogene = 32431\n",
      "ncRNA = 4565\n",
      "Can = 525753\n"
     ]
    }
   ],
   "source": [
    "#Specify protein files to use\n",
    "prot_files = ['proteins.fa']#, 'sixframes.fa']\n",
    "\n",
    "proteins = {}\n",
    "for prot_file in prot_files:\n",
    "    for record in SeqIO.parse(prot_file, 'fasta'):\n",
    "        prot_type = str(record.id).replace('ENS', 'Can_ENS').split('_')[0]\n",
    "        try:\n",
    "            proteins[prot_type].extend(trypsin_digestion(str(record.seq)))\n",
    "        except KeyError:\n",
    "            proteins[prot_type] = trypsin_digestion(str(record.seq))\n",
    "\n",
    "#unify\n",
    "for prot_type in proteins.keys():\n",
    "    proteins[prot_type] = set(proteins[prot_type])\n",
    "\n",
    "print('Number of tryptic peptides per type:\\n', '\\n'.join([k+' = '+str(len(items)) for k,items in proteins.items()]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "1d9241bd",
   "metadata": {
    "scrolled": true,
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABH4AAAJCCAYAAAC23QD1AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAgAElEQVR4nOzdebhVZcH38d9mHg7IqIAyOKQmpj4+DpWkaIqgQomVA+KURqlZvVbqo6mk2WD1SvlYDqWAqWVpKoU4ZZqFmoWUigMGMinOcEDm/f5R7tcjlBTngN5+Ptd1rva+17rXupd/cJ2+Z629K9VqNQAAAACUp9mGXgAAAAAATUP4AQAAACiU8AMAAABQKOEHAAAAoFDCDwAAAEChhB8AAACAQrVY3yfs0q5dtXenTg3GWvbqleVz5zbaOVr26tVoxwLePRrz3yEAAKBxzVi0KBtvvPGGXsbb1lNPPVWtVqur3eCz3sNP706dcuvxJzQY6/nV0Zl39jmNdo6eXx3daMcC3j0a898hAACgcQ386XW59NJLN/Qy3rb23nvv19Y07lEvAAAAgEIJPwAAAACFEn4AAAAACiX8AAAAABRK+AEAAAAolPADAAAAUCjhBwAAAKBQwg8AAABAoYQfAAAAgEIJPwAAAACFEn4AAAAACiX8AAAAABRK+AEAAAAolPADAAAAUCjhBwAAAKBQwg8AAABAoYQfAAAAgEIJPwAAAACFEn4AAAAACiX8AAAAABRK+AEAAAAolPADAAAAUCjhBwAAAKBQwg8AAABAoYQfAAAAgEIJPwAAAACFEn4AAAAACiX8AAAAABRK+AEAAAAolPADAAAAUCjhBwAAAKBQwg8AAABAoYQfAAAAgEIJPwAAAACFEn4AAAAACiX8AAAAABRK+AEAAAAolPADAAAAUCjhBwAAAKBQwg8AAABAoYQfAAAAgEIJPwAAAACFEn4AAAAACiX8AAAAABRK+AEAAAAolPADAAAAUCjhBwAAAKBQwg8AAABAoYQfAAAAgEIJPwAAAACFEn4AAAAACiX8AAAAABRK+AEAAAAolPADAAAAUCjhBwAAAKBQwg8AAABAoVps6AUAAO8ON/zlL7ns/sl56oUXUte6dfpvsklOGfCh7N6nz4ZeGgBQuM9//vN59NFH07x58yRJ9+7dM27cuDz66KP58Y9/nCeeeCLNmjXLTjvtlM9+9rPp2rVrkuSqq67K1VdfnZYtW9aO9aMf/Si9evXKc889l2OOOabBeZYsWZLPfOYz+cQnPpEpU6bk//yf/5PWrVvXtn/uc5/L4MGDkyQ//OEPc9ddd2XRokXp0KFDDjrooBx55JGNfu3CDwDQ5C6d/IdcfN99+eYBB2bgllumZfPm+c30pzLpiceFHwBgvfjc5z6XAw88sMHYwoULc9BBB2XXXXdN8+bNM2bMmHzzm9/Mt771rdo+e++9d84888zVjrfJJptk4sSJtffz5s3LkUcemT333LM21rVr11x//fVrXM8BBxyQo48+Om3bts3zzz+fL33pS+nTp0+D+Y1B+AEAmtSCJUty4d135/8O+0gOeO97a+ODtt4mg7beJn+eMydnT7o1T77wQtq0bJkDtn1vzh00KK3+8Re5Xud9Nd844IBcOnlyXlq8OB/dfvtcMHhIKpXKhrokAKAQu+++e4P3Bx98cD7/+c//R8e67bbbssMOO6RHjx5rtX+fN/3xq1mzZpkzZ85/dO5/RfgBAJrUQ7NnZ+mKFRmy7bZr3N68WSXnDto/O/bqlXkLFmTENddk7B8fzAm7v7+2zx1PPplff/L4LFy6NIOvuDyD3rN19t5qq/V1CQBAAS6//PJcdtll6d27d44//vjstNNOq+0zderU9OvXr8HYH/7whwwbNixdunTJwQcfnI985CNrPP5tt92WkSNHNhh75ZVXMnz48LRu3ToDBgzIcccdl7Zt29a2X3PNNRk/fnyWLFmSnj17Zt999133C30T4QcAaFIvv/ZaurRrlxbN1vydEjv07FV73btTp4z8753zh5kzG4Sfkz64RzZq0yYbtWmTPfr2y1+fe1b4AQDW2qc+9an069cvLVq0yF133ZX/+Z//yeWXX55NN920ts/06dMzbty4nH/++bWxgQMH5qCDDkrnzp3z2GOP5ZxzzkldXV0+/OEPNzj+1KlT89JLL2WvvfaqjfXp0yeXX355+vTpk+eeey7f+MY3cskll+TUU0+t7XPEEUfk8MMPz1NPPZXf/e53ad++faNfu2/1AgCaVOe2bfPS4sVZsWrVGrdPf/HFHHXdtdnxu9/J1t/8Rr5+1115afHiBvtsXFdXe922ZcssXrasSdcMAJRlu+22S7t27dKqVasMHjw422+/fe6///7a9jlz5uT000/PySefnB122KE23q9fv3Tr1i3NmzfP9ttvn0MOOSS//e1vVzv+pEmTsueeeza4m6dLly7p169fmjVrlp49e2bUqFG55557VptbqVTynve8J61bt86VV17ZyFcu/AAATey/N9ssrVu0yK3Tpq1x+xm//lW27No19510cp447fScsc8+qa7nNQIA7y6VSiXV6t9/43j22Wdz6qmnZuTIkRk0aNBbznuzpUuX5re//W3233//tzzv6+dck5UrV2bu3LlveYx/l/ADADSpjm3a5EsDB+Z/bp2YidOmZfHy5Vm+cmXueurJnHfH7alftiwdWrdO+1at8uQLL2TsHx/a0EsGAApSX1+fBx54IMuWLcvKlStz++23Z+rUqdl1113z/PPP59RTT81HP/rRDBs2bLW5v/vd77Jw4cJUq9U89thjueGGG/LBD36wwT733ntv6urq8l//9V8NxqdMmZLnnnsu1Wo18+fPz+WXX5499tgjSbJq1arcfPPNDY79y1/+MjvvvHOjX7/P+AEAmtyo938g3dvXZczv7s3Jv7wxda1aZYeePXPKgA9lv/dsnS/9akIu+f3vs32PHhnWf7vcN2PGhl4yAFCIFStW5Mc//nGeeeaZNGvWLH369Ml5552XPn36ZOzYsZk7d27Gjh2bsWPH1ua8/jXtv/nNb3LhhRdm2bJl6d69ew477LAMHjy4wfEnTZqUQYMGrXY30BNPPJGvfe1rWbhwYTp27JgBAwbk+OOPr23/3e9+lyuuuCLLly9Pt27dcvDBB2f48OGNfv2Vf3WbUVPYsVev6q3Hn9BgrOdXR2fe2ec02jl6fnV0ox0LePdozH+HAACAxjXwp9fl0ksv3dDLeNvae++9F1er1dU+HdqjXgAAAACFEn4AAAAACiX8AAAAABRK+AEAAAAolPADAAAAUCjhBwAAAKBQwg8AAABAoYQfAAAAgEIJPwAAAACFEn4AAAAACiX8AAAAABRK+AEAAAAolPADAAAAUCjhBwAAAKBQwg8AAABAoYQfAAAAgEIJPwAAAACFEn4AAAAACiX8AAAAABRK+AEAAAAolPADAAAAUCjhBwAAAKBQwg8AAABAoYQfAAAAgEIJPwAAAACFEn4AAAAACiX8AAAAABRK+AEAAAAolPADAAAAUCjhBwAAAKBQwg8AAABAoYQfAAAAgEIJPwAAAACFEn4AAAAACiX8AAAAABRK+AEAAAAolPADAAAAUCjhBwAAAKBQwg8AAABAoYQfAAAAgEIJPwAAAACFEn4AAAAACiX8AAAAABRK+AEAAAAolPADAAAAUCjhBwAAAKBQwg8AAABAoYQfAAAAgEIJPwAAAACFEn4AAAAACiX8AAAAABRK+AEAAAAolPADAAAAUCjhBwAAAKBQwg8AAABAoYQfAAAAgEIJPwAAAACFEn4AAAAACtVifZ/w+Y2Snl8dvdr4ZUOa5ZwPnNMo5xj9h9WP31jHBsq1pn+bAACAt4dm1/9sQy/hHWm9hx8AAACAf1f79u0zcODADb2MdxyPegEAAAAUSvgBAAAAKJTwAwAAAFAo4QcAAACgUMIPAAAAQKGEHwAAAIBCCT8AAAAAhRJ+AAAAAAol/AAAAAAUSvgBAAAAKJTwAwAAAFAo4QcAAACgUMIPAAAAQKGEHwAAAIBCCT8AAAAAhRJ+AAAAAAol/AAAAAAUSvgBAAAAKJTwAwAAAFAo4QcAAACgUMIPAAAAQKGEHwAAAIBCCT8AAAAAhRJ+AAAAAAol/AAAAAAUSvgBAAAAKJTwAwAAAFAo4QcAAACgUMIPAAAAQKGEHwAAAIBCCT8AAAAAhRJ+AAAAAAol/AAAAAAUSvgBAAAAKJTwAwAAAFAo4QcAAACgUMIPAAAAQKGEHwAAAIBCCT8AAAAAhRJ+AAAAAAol/AAAAAAUSvgBAAAAKJTwAwAAAFAo4QcAAACgUMIPAAAAQKGEHwAAAIBCCT8AAAAAhRJ+AAAAAAol/AAAAAAUSvgBAAAAKJTwAwAAAFAo4QcAAACgUMIPAAAAQKGEHwAAAIBCCT8AAAAAhRJ+AAAAAAol/AAA693AgQPTpk2b1NXVpVu3bhk+fHjmzZtX237uueemUqnk+uuvr42tWLEilUolM2bMaHCs1/d94IEH1tfyAYACPfnkk2nTpk2OPPLIJMmMGTNSqVRSV1dX+znvvPMazPnTn/6UPffcM3V1ddlkk00yZsyY2rZ+/fqlbdu2tbmDBg1ar9fzOuEHANggLr744tTX1+epp55KfX19vvjFLzbY3qVLl5x99tlZuXLlPz1GtVrN+PHj06VLl4wdO7aplwwAFOykk07Krrvuutr4K6+8kvr6+tTX1+crX/lKbfyFF17I4MGDM2rUqLz44ot56qmnVos7t9xyS23ubbfd1uTXsCbCDwCwTvr165dvf/vb2WGHHbLRRhvl0EMPzZIlS5IkN910U3baaad07NgxW265ZW699dbV5nfq1Ckf/ehHM2XKlAbjgwcPTqtWrXL11Vf/03Pfe++9mTt3bsaMGZPrrrsuy5Yta9yLAwDeFa677rp06tQpH/7wh9d6zne/+93sv//+GTFiRFq3bp0OHTrkve99bxOu8j8j/AAA6+xnP/tZbr311vztb3/L1KlTc9VVV+WBBx7IUUcdlQsvvDCvvPJK7rnnnvTr12+1uS+++GJuuOGGbLXVVg3GK5VKzjvvvIwePTrLly9f43nHjh2boUOH5tBDD02STJgwodGvDQAo24IFC3L22WfnO9/5zhq39+3bN5tttlmOPfbYvPDCC7XxyZMnp0uXLvngBz+YjTfeOEOHDs0zzzzTYO6IESPSvXv3DBo0KA8//HCTXsc/I/wAAOvslFNOSa9evdKlS5cMHTo0U6ZMyY9+9KMcd9xx2W+//dKsWbNsuumm2XbbbRvM2WijjdKtW7e88MIL+f73v7/acYcNG5bu3bvniiuuWG3b4sWLc/311+eII45Iy5Yt87GPfczjXgDAv+0rX/lKPvnJT6Z3794Nxrt165YHH3wwM2fOzEMPPZSFCxdmxIgRte2zZ8/O2LFjM2bMmDzzzDPZfPPNc/jhh9e2/+QnP8mMGTMyc+bM7L333tl///3zyiuvrLfrep3wAwCssx49etRet2vXLvX19Zk1a1a23HLLfzrne9/7Xl599dVMnTo1L7/8cmbPnr3G/c4///x87Wtfqz0+9robb7wxLVq0yAEHHJDk739RmzhxYp5//vlGuCIA4N1gypQpueOOO/KFL3xhtW11dXXZZZdd0qJFi2yyySa5+OKLc9ttt2XBggVJkrZt2+bggw/OrrvumjZt2uScc87J73//+7z66qtJkj322CNt27ZNu3btcsYZZ6RTp06599571+v1JUmL9X5GAOBdoXfv3pk+ffpb7ve+970vZ511Vk466aT86U9/SqVSabB9v/32y1ZbbZVLLrmkwfjYsWNTX1+fPn36JPn7Bz0vX7481157bU455ZTGuxAAoFh33313ZsyYUft9or6+PitXrsyjjz6aP/3pTw32ff13lGq1miTZYYcdGvze8ubtb1apVP7ptqbkjh8AoEl88pOfzJVXXpk777wzq1atypw5czJt2rQ17nv00Udn/vz5ufnmm9e4/Wtf+1q+9a1v1d7PmTMnd955ZyZMmJApU6ZkypQpefjhh3Paaad53AsAWGuf+tSnMn369NrvE5/+9Kdz4IEHZtKkSbn//vvz+OOPZ9WqVXnxxRdzyimnZODAgdloo42SJMcee2xuvPHGTJkyJcuXL895552XAQMGpFOnTnnmmWdy3333ZdmyZVmyZEkuvPDCvPDCC9ljjz3W+zUKPwBAk9htt91y5ZVX5gtf+EI22mij7LXXXpk5c+Ya923VqlVOOeWUnHfeeWvcvscee2S33XarvR8/fnx22mmnDBo0KD169Kj9nHLKKZk6dWr++te/Nsk1AQBladeuXYPfJerq6tKmTZt07949Tz/9dAYPHpwOHTpk++23T+vWrXPttdfW5u6zzz654IILcuCBB2bjjTfOU089lWuuuSZJsnDhwnzmM59J586ds+mmm+bWW2/NxIkT07Vr1/V+jZX1fZtRr/f2qs59bO5q46P/MDrnfOCcRjnH6D+MXm2ssY4NAAAArH+77LJL/vjHP27oZbxtVSqVxdVqtf2bx93xAwAAAFAo4QcAAACgUMIPAAAAQKGEHwAAAIBCCT8AAAAAhRJ+AAAAAAol/AAAAAAUSvgBAAAAKJTwAwAAAFAo4QcAAACgUMIPAAAAQKGEHwAAAIBCCT8AAAAAhRJ+AAAAAAol/AAAAAAUSvgBAAAAKJTwAwAAAFAo4QcAAACgUMIPAAAAQKGEHwAAAIBCCT8AAAAAhRJ+AAAAAAol/AAAAAAUSvgBAAAAKJTwAwAAAFAo4QcAAACgUMIPAAAAQKGEHwAAAIBCCT8AAAAAhRJ+AAAAAAol/AAAAAAUSvgBAAAAKJTwAwAAAFAo4QcAAACgUMIPAAAAQKGEHwAAAIBCCT8AAAAAhRJ+AAAAAAol/AAAAAAUSvgBAAAAKJTwAwAAAFAo4QcAAACgUMIPAAAAQKGEHwAAAIBCCT8AAAAAhRJ+AAAAAAol/AAAAAAUSvgBAAAAKJTwAwAAAFAo4QcAAACgUMIPAAAAQKGEHwAAAIBCCT8AAAAAhRJ+AAAAAAol/AAAAAAUSvgBAAAAKJTwAwAAAFAo4QcAAACgUMIPAAAAQKFarO8T9lqxMrnlc6uNnzN0zP8fHzpmnc5xzgfOWX1wDed821jH6wUAAIDSvfrqq5kwYULtfV1d3QZczTvHeg8/AAAAAP+uZs2aNYg9AwcO3HCLeQfxqBcAAABAoYQfAAAAgEIJPwAAAACFEn4AAAAACvUfhZ9KpdKpUqn8V6VSadfYCwIAAACgcbxl+KlUKl+qVCrD3/B+cJJZSR5KMqtSqezehOsDAAAA4D+0Nnf8fDLJX9/w/nv/+OmQ5LtJvt4E6wIAAABgHa1N+OlZrVafSJJKpbJVkr5Jvl6tVhcl+XaSHZpwfQAAAAD8h9Ym/CyuVCod//F6QJKp1Wq1/h/vVyVp0SQrAwAAAGCdrE34mZjkskqlMizJF5P84g3bdszfP+8HAAAAgLeZtQk/X0iyOMnXkvwhyf99w7bBSa5rgnUBAAAAsI7W5jGtA6rV6nFr2lCtVs9v5PUAAAAA0EjW5o6fS5t8FQAAAAA0urUJP5UmXwUAAAAAjW5tHvVqXqlU9s6/CEDVavWuxlsSAAAAAI1hbcJP6yQ/yj8PP9UkWzTaigAAAABoFGsTfhZVq1VhBwAAAOAdZm0+4wcAAACAdyAf7gwAAABQqLUJP1v/q42VSuW/G2ktAAAAADSitQk/j7/xTaVSefJN23/TeMsBAAAAoLH8J496dXuL7QAAAAC8DaxN+Kn+m+8BAAAAeBvwrV4AAAAAhWqxFvu0q1Qq97zhfYc3vK8kadv4ywIAAABgXa1N+Pnkm97/6E3vr2iktQAAAADQiNYm/Pw1ydJqtfrXJKlUKhsnuShJ/ySTk5zadMsDAAAA4D+1Np/xc1GSHm94f1mS9/zjf/sn+VYTrAsAAACAdbQ2d/y8N8m9SVKpVDolOTBJ/2q1+kSlUrk5ye+TnNh0SwQAAADgP7E2d/y0SLLsH6/fn2RetVp9Ikmq1eqsJJ2aaG0AAAAArIO1CT+PJPn4P14fluSO1zdUKpVNk7zaBOsCAAAAYB2tzaNepyW5pVKp/DDJyiQD3rDt0CT3NcXCAAAAAFg3bxl+qtXq7yqVSp8kWyd5olqtLnzD5l8lua6pFgcAAADAf25t7vjJP2LPQ2sYf7zRVwQAAABAo1ibz/gBAAAA4B1I+AEAAAAolPADAAAAUCjhBwAAAKBQwg8AAABAoYQfAAAAgEIJPwAAAACFEn4AAAAACiX8AAAAABRK+AEAAAAolPADAAAAUCjhBwAAAKBQwg8AAABAoYQfAAAAgEIJPwAAAACFEn4AAAAACiX8AAAAABRK+AEAAAAolPADAAAAUCjhBwAAAKBQwg8AAABAoYQfAOBt4e67785mm222oZcBAFAU4QcAAAB41zjyyCPTs2fPdOzYMVtvvXWuuOKKJMnkyZOz3377pUuXLunevXs+/vGPZ968eavNX7ZsWbbddtvV/mC19957p3v37unYsWN23HHH3HTTTbVt8+bNy7Bhw9KrV69UKpXMmDGjSa/xjYQfAAAA4F3jjDPOyIwZM7JgwYLcfPPNOeuss/LQQw/l5Zdfzqc+9anMmDEjM2fOTIcOHXLssceuNv/CCy/MxhtvvNr4mDFjMm/evCxYsCCXXXZZjjzyyFo4atasWQYPHpxf/OIXTX59byb8AMC7SL9+/fL1r3892223XTp37pxjjz02S5YsSZJMmDAhO+20Uzp16pQPfvCDmTp1am3eN7/5zWy66abp0KFDttlmm9x5551JkmOOOSZnnXVWbb83P641d+7cHHLIIenevXs233zzfO9736tte+2113LMMcekc+fO2W677fLggw82WOtjjz2WgQMHplOnTunfv39uvvnm2rYXX3wxQ4cOTceOHbPrrrvmrLPOyoABA2rbp02bVvuL3TbbbJOf/exntW3HHHNMTjrppBx44IHp0KFDdt9990yfPn2t5gIA73z9+/dP69atkySVSiWVSiXTp0/PkCFD8vGPfzwdO3ZMu3btcvLJJ+e+++5rMPdvf/tbrr766pxxxhmrHXeHHXZIixYtasddvnx5Zs2alSTZZJNNcuKJJ2bXXXdt4qtbnfADAO8yP/nJTzJp0qRMnz49TzzxRM4///z86U9/ynHHHZdLL700L774YkaNGpVhw4Zl6dKlefzxx3PxxRfnwQcfzMKFCzNp0qT069fvLc+zatWqDB06NDvuuGPmzJmTO++8MxdddFEmTZqUJBk9enSmT5+e6dOnZ9KkSRk7dmxt7vLlyzN06NAMGjQo8+fPz/e///2MGDEijz/+eJLkpJNOSvv27fPss89m7NixDeYuWrQo++23X4444ojMnz8/1157bU488cQ88sgjtX2uvfbanHPOOXn55Zez1VZb5cwzz1zruQDAO9+JJ56Ydu3aZdttt03Pnj1zwAEHrLbPPffck/79+zcY++xnP5sLLrggbdu2XeNxDzrooLRp0ya77757Bg4cmF122aVJ1v/vEH4A4F3m5JNPTu/evdOlS5eceeaZufbaa3P55Zdn1KhR2X333dO8efMcffTRad26dSZPnpzmzZtn6dKlefTRR7N8+fL069cvW2655Vue58EHH8zzzz+fs88+O61atcoWW2yRE044Idddd12S5Gc/+1nOPPPMdOnSJb17984pp5xSmzt58uTU19fn9NNPT6tWrbLPPvvkoIMOyrXXXpuVK1fmF7/4RUaPHp127dplu+22y9FHH12bO2HChPTr1y/HHntsWrRokZ133jmHHHJIfv7zn9f2GT58eHbbbbe0aNEiI0aMyJQpU9Z6LgDwznfJJZdk4cKFuffeezN8+PDaHUCvmzp1ar761a/mwgsvrI3deOONWbFiRQ4++OB/etwJEyZk4cKF+fWvf539998/zZpt+Oyy4VcAAKxXvXv3rr3u27dv5s6dm5kzZ+Y73/lOOnXqVPuZNWtW5s6dm6222ioXXXRRzj333Gy88cY57LDDMnfu3Lc8z8yZMzN37twGx7zgggvy3HPPJfn7Y2BvXsvrXt/2xl+W+vbtmzlz5uT555/PihUrGsx94+uZM2fm/vvvb3Den/zkJ3n22Wdr+/To0aP2ul27dqmvr1/ruQBAGZo3b54BAwZk9uzZ+cEPflAbf+qppzJkyJCMGTMmH/rQh5L8/a7gL3/5y/n+97//lsdt2bJlhgwZkkmTJjV4VH1DabGhFwAArF+vP2ueJM8880x69eqV3r1758wzz6w98vRmRxxxRI444ogsWLAgo0aNymmnnZbx48enffv2Wbx4cW2/NwaS3r17Z/PNN8+TTz65xmP27Nkzs2bNqt1C/cwzz9S29erVK7NmzcqqVatq8eeZZ57J1ltvne7du6dFixaZPXt2tt5669WuqXfv3tlrr71y++23/7v/adZpLgDwzrRixYra5/3NnDkz++67b77yla9k5MiRtX2efPLJzJgxoxaCli1blldffTU9evTI5MmT1/gY/BuPuyG54wcA3mX+93//N7Nnz85LL72UCy64IIceemhOOOGE/PCHP8z999+farWaRYsW5Ve/+lUWLlyYxx9/PHfddVeWLl2aNm3apG3btmnevHmSZKeddsqvf/3rvPTSS3n22Wdz0UUX1c6z2267pWPHjvnmN7+Z1157LStXrsxf//rX2oc4f+ITn8jXv/71vPzyy5k9e3aDv6Dtvvvuad++fb71rW9l+fLlufvuu3PLLbfksMMOS/PmzTN8+PCce+65Wbx4caZNm5Zx48bV5h500EF54oknMn78+CxfvjzLly/Pgw8+mMcee+wt/9usy1wA4O1v/vz5ue6661JfX5+VK1dm0qRJufbaa7PPPvtkzpw52WeffXLSSSfl05/+dIN522+/fWbNmpUpU6ZkypQpueKKK7LJJptkypQp6d27d6ZNm5aJEyfmtddey/Lly3P11VfnnnvuyV577VU7xpIlS7J06dIkydKlS2tfsNHUhB8AeJc54ogjMmjQoGyxxWsTw10AACAASURBVBbZYostctZZZ2WXXXbJ5ZdfnpNPPjmdO3fOVlttlauuuirJ338xOf3009OtW7f06NEj8+fPzwUXXJAkGTlyZHbcccf069cvgwYNyqGHHlo7T/PmzXPLLbdkypQp2XzzzdOtW7ccf/zxefXVV5Mk55xzTvr27ZvNN988gwYNavBXtVatWuXmm2/OxIkT061bt5x44okZN25ctt122yTJxRdfXPsr28iRI3P44YfXns3v0KFDbrvttlx33XXp1atXevTokdNOO632i9a/si5zAYC3v0qlkh/84AfZbLPN0rlz53zxi1/MRRddlI985CO54oor8vTTT2f06NGpq6ur/SRJixYt0qNHj9pPly5d0qxZs/To0SPNmzdPtVqtPRbfvXv3jBkzJj/96U+z8847187dtm3b2vG23Xbbf/oB0Y1+zdVqdb2c6HW7vGeT6h+/e9jqG4aOSW753P9/3dheP/bbUVNcLwCsQb9+/XLFFVdk33333dBLaVSnnXZa7Ru+AIAybbPNNrn00ktr7wcOHLjhFvM2VKlUFler1fZvHnfHDwDwjjNt2rRMnTo11Wo1DzzwQH70ox/9y2/YAAB4t/LhzgDAO87ChQtz+OGHZ+7cudl4441z6qmn5iMf+ciGXhYAwNuO8AMA7yIzZszY0EtoFLvuumueeuqpDb0MAIC3PY96AQAAABRK+AEAAAAolPADAAAAUCjhBwAAAKBQwg8AAABAoYQfAAAAgEIJPwAAAACFEn4AAAAACiX8AAAAABRK+AEAAAAolPADAAAAUCjhBwAAAKBQwg8AAABAoYQfAAAAgEIJPwAAAACFEn4AAAAACiX8AAAAABRK+AEAAAAolPADAAAAUCjhBwAAAKBQwg8AAABAoYQfAAAAgEIJPwAAAACFEn4AAAAACiX8AAAAABRK+AEAAAAolPADAAAAUCjhBwAAAKBQwg8AAABAoYQfAAAAgEIJPwAAAACFEn4AAAAACiX8AAAAABRK+AEAAAAolPADAAAAUCjhBwAAAKBQwg8AAABAoYQfAAAAgEIJPwAAAACFEn4AAAAACiX8AAAAABRK+AEAAAAolPADAAAAUCjhBwAAAKBQwg8AAABAoYQfAAAAgEIJPwAAAACFEn4AAAAACiX8AAAAABRK+AEAAAAolPADAAAAUCjhBwAAAKBQwg8AAABAoYQfAAAAgEIJPwAAAACFarHez7hR75yx/PjVhr+eJEPHJEnOuOEv63SKrw9/3+qD/zj229G6Xi/QONb4bwcAAPC2sGrVqtTX19feT5gwodGOXVdX12jHertZ/+EHAAAA4N/UrFmzJgs0AwcObJLjvh141AsAAACgUMIPAAAAQKGEHwAAAIBCCT8AAAAAhRJ+AAAAAAol/AAAAAAUSvgBAAAAKJTwAwAAAFAo4QcAAACgUMIPAAAAQKGEHwAAAIBCCT8AAAAAhRJ+AAAAAAol/AAAAAAUSvgBAAAAKJTwAwAAAFAo4QcAAACgUMIPAAAAQKGEHwAAAIBCCT8AAAAAhRJ+AAAAAAol/AAAAAAUSvgBAAAAKJTwAwAAAFAo4QcAAACgUMIPAAAAQKGEHwAAAIBCCT8AAAAAhRJ+AAAAAAol/AAAAAAUSvgBAAAAKJTwAwAAAFAo4QcAAACgUMIPAAAAQKGEHwAAAIBCCT8AAAAAhRJ+AAAAAAol/AAAAAAUSvgBAAAAKJTwAwAAAFAo4QcAAACgUMIPAAAAQKGEHwAAAIBCCT8AAAAAhRJ+AAAAAAol/AAAAAAUSvgBAAAAKJTwAwAAAFAo4QcAAACgUMIPAAAAQKGEHwAAAIBCCT8AAAAAhRJ+AAAAAAol/AAAAAAUSvgBAAAAKJTwAwD8R/r165c77rhjQy8DAIB/QfgBADaIq666Ks2bN09dXV06duyYHXfcMRMmTKhtnzFjRiqVSg488MAG84488sice+65Dcb+9re/pVmzZjnxxBPXx9IBgHeQG2+8MaNGjcqgQYPyjW98ozZ+++23Z8iQIRkyZEjq6urSrl27VCqVPPTQQ0mSarWa0047LV27dk3Xrl3z5S9/OdVqtTZ/7733Tvfu3Wu/x9x0000NznvNNdekb9++ad++fT760Y/mpZdeqm3r379/6urqaj8tWrTI0KFDm+T6hR8AYIP5wAc+kPr6+rzyyis58cQTc9hhh+WVV15psM/kyZNz3333/cvjjBs3Lp07d851112XpUuXNuWSAYB3mK5du2bkyJEZMmRIg/H99tsvEydOzMSJE1NfX59LLrkkW2yxRXbeeeckyWWXXZZf/vKXefjhhzN16tRMmDAhl156aW3+mDFjMm/evCxYsCCXXXZZjjzyyMybNy9J8sgjj2TUqFEZP358nnvuubRr167BH6geeeSR1NfXp76+PgsXLkyfPn3y8Y9/vEmuX/gBANbJueeem0984hM56qij0qFDh/Tv3z9//OMfa9tnzZqV4cOHp3v37unatWtOPvnk1Y7RrFmzjBw5MosWLcqTTz7ZYNuXv/zlnHXWWf9yDePGjcv555+fli1b5pZbbmmcCwMAirDnnntmwIAB6dix47/cb+zYsTnqqKNSqVRq70899dRsttlm2XTTTXPqqafmqquuqu2/ww47pEWLFkmSSqWS5cuXZ9asWUmSn/zkJxk6dGj23HPP1NXV5bzzzssNN9yQhQsXrnbee+65J/Pnz88hhxzSSFfckPADAKyzm2++uXa3zrBhw2pxZ+XKlTnooIPSt2/fzJgxI3PmzMlhhx222vyVK1fmyiuvTMuWLdO3b98G20466aQ88cQT//TzhO69997Mnj07hx12WD7xiU9k3LhxjX+BAEDRZs6cmXvuuSdHHXVUbeyRRx7JjjvuWHu/44475pFHHmkw76CDDkqbNm2y++67Z+DAgdlll13WOHfLLbdMq1at8sQTT6x27rFjx+ZjH/tY2rdv39iXlSRp0SRHBQDeVQYMGJADDjggSTJy5MhcdNFFSZIHHnggc+fOzYUXXlj7i9iAAQNq8yZPnpxOnTpl0aJFadGiRa6++upsvPHGDY7dpk2bnHnmmTnrrLOy7777rnbusWPHZsiQIencuXOOOOKI7Lnnnpk/f/5qxwEA+GfGjRuXD33oQ9l8881rY/X19dloo41q7zfaaKPU19enWq3W7gqaMGFCli9fnjvuuCPTpk1Ls2bN1jj39flvvuNn8eLF+fnPf56bb765qS7NHT8AwLrr0aNH7XW7du2yZMmSrFixIrNmzUrfvn1r0efN3v/+9+eVV17Jyy+/nGHDhuXee+9d434nnHBCnnvuudUe43rttddy/fXXZ8SIEUn+/plBffr0yTXXXNNIVwYAvBuMGzcuRx99dIOxurq6LFiwoPZ+wYIFqaurq0Wf17Vs2TJDhgzJpEmTagHnzXNfn9+hQ4cGYzfccEO6dOmSvfbaqzEvpwHhBwBoMr17984zzzyTFStW/Mv96urqcskll2T8+PH585//vNr2li1b5pxzzslXvvKVBt+mceONN2bBggU58cQT06NHj/To0SNz5szxuBcAsNb+8pe/ZO7cufnYxz7WYLx///55+OGHa+8ffvjh9O/f/58eZ8WKFZk+ffoa5z799NNZunRptt566wZz3vy5Qk1B+AEAmsxuu+2Wnj175vTTT8+iRYuyZMmSf/oNXV27ds3xxx+fr371q2vcPnLkyCxdujS33nprbWzs2LE57rjj8pe//CVTpkzJlClTct9992XKlCn5y1/+0iTXBAC8s6xcuTLLli3LqlWrsmrVqixbtiwrV66sbZ80aVIOOeSQ1e7GOeqoo/Ld7343c+bMydy5c/Od73wnxxxzTJJk2rRpmThxYl577bUsX748V199de65557anTsjRozILbfcknvvvTeLFi3K2WefneHDhzc4x+zZs/Ob3/xmtTuNGpvP+AEAmkzz5s1zyy235JRTTkmfPn1SqVRyxBFHZI899ljj/p///Oez5ZZbZurUqat980bz5s0zevToHHrooUmSOXPm5M4778yf//znBo+a9ejRI4MHD87YsWPz7W9/u+kuDgB4Rxg/fnzGjh1be3/77bfn6KOPzjHHHJNly5bl7rvvzk033bTavFGjRuXpp5/O+973viTJ8ccfn1GjRiVJqtVqzj333Dz66KNp3rx53vOe9+SnP/1p7avg+/fvnx/+8IcZMWJEXnzxxey777658sorV1vXBz7wgWy55ZZNdelJksobb5deH3bZZZfqfv9z5WrjXx/+vtrrM25Yt7/QvfFY7wTrer1A43in/dsBAADvJttss00uvfTSJjn2wIEDm+S461OlUllcrVZX+2owj3oBAAAAFEr4AQAAACiU8AMAAABQKOEHAAAAoFDCDwAAAEChhB8AAACAQgk/AAAAAIUSfgAAAAAKJfwAAAAAFEr4AQAAACiU8AMAAABQKOEHAAAAoFDCDwAAAEChhB8AAACAQgk/AAAAAIUSfgAAAAAKJfwAAAAAFEr4AQAAACiU8AMAAABQKOEHAAAAoFDCDwAAAEChhB8AAACAQgk/AAAAAIUSfgAAAAAKJfwAAAAAFEr4AQAAACiU8AMAAABQKOEHAAAAoFDCDwAAAEChhB8AAACAQgk/AAAAAIUSfgAAAAAKJfwAAAAAFEr4AQAAACiU8AMAAABQKOEHAAAAoFDCDwAAAEChhB8AAACAQgk/AAAAAIUSfgAAAAAKJfwAAAAAFEr4AQAAACiU8AMAAABQKOEHAAAAoFDCDwAAAEChhB8AAACAQgk/AAAAAIUSfgAAAAAKJfwAAAAAFEr4AQAAACiU8AMAAABQKOEHAAAAoFDCDwAAAEChhB8AAACAQgk/AAAAAIUSfgAAAAAKJfwAAAAAFKrF+j7h448/nvlnH7fa+B++1772+ukXFq3TOd54rHeCdb1eoHG80/7tAAAAeCvrPfwkyRbd/vX/ufp/7d17cNXlncfxz5MbnJAbgZBQCZQIWDdpoTFUKiCppsUQO2xIrUiRax1Xisvi1q0VdTAj1HZaK5PaGaGI1Ah2VpNtCNCKVlbBFW9ASHAwRCSElFu4JVCOMXn2j4QA5uJU8zu/X3Ler5kz5HeeJN9PIDM5+fI839/nrfc2wfb1AgAAdKfy8nIlJCRo4MCBCg0NdTsOAMAhISHOHVraunWrY5/bbcZaG9CCGRkZ9t133w1oTaC7TZ06VT/84Q81bdo0+Xw+t+MAABDUwsPDNXToUB0/flw/+MEPNHfuXE2cONHtWACAbpaRkSH6CZ0zxpy31rbbWcKMH+AL2LRpkx555BElJiZq3rx5euONN9yOBABA0PL5fKqqqtKGDRtkjNGtt96qlJQULV26VAcOHHA7HgAArqLx8wW98MIL8vv9bseAS3iBCQCA90yaNEmrV6/WkSNHlJ+frzfffFOjRo1SZmam29EAAHBNj2j8WGt19OjRtusjR46oqKhIH374oeO1P/roow4fixYtUllZmT766CPHM8C7eIEJAID3+Hw+zZw5Uy+//LIOHDigW265xe1IAAC4xvMzfrZv367c3FzV1dVpzJgxWrlypXJycjR48GBVVlZq5cqVmjFjhmN5Q0JCZIxRZ39Pxhg1NTU5Vh/eFBMTo7Nnz3a4VlNTo8LCQj3wwAMBTgUAQHBKS0tTeXm52zEAAA5jxk/XeuyMn8WLFys/P1/19fWaM2eOsrOztXbtWu3cuVPFxcVatmyZo/Vnz56tCRMmqKysTM3NzW2PxMRE/f3vf6fpE6SGDh3a6dqQIUNo+gAAEEA0fQAA6Jznd/zExcXp9OnTkqTm5mb17dtXn3zySdt6bGyszpw50+05L7d9+3bde++9yszMVH5+vqKiojR48GDt3r1bgwYNcrQ2AAAAAABgx8/n6bE7fuLj47Vr1y5JavsHrqqqkiQdOHBAcXFxjmcYP3683nnnHQ0ePFjp6ekqLCx0vCZ6nubm5rYmJQAA8Aa/36/Q0FC3YwAA4JowtwN8nsWLF2vChAlKS0tTU1OTHn30UWVnZ+v73/++Nm7cqLlz5wYkR2hoqO6//35Nnz5dixcv1oULFxQS4vm+GRxSW1ure+65R1VVVfrZz36mUaNGaerUqTp+/LjS09NVVFSk5ORkt2MCABAUqqurO127cOFCp7MaAQAIBp4/6iVJb7/9tj7++GNlZ2crOjpaq1atUllZmcaOHatZs2Y5lBToXF5enkaMGKGQkBA98cQTeuyxxzRr1iw1NjbqvvvuU1hYmNatW+d2TAAAggI34wCA4MBRr651dtSrRzR+3FZXV6eXXnpJFRUVqq+vV3R0tNLS0jRt2jQNGDDA7XhwQUJCgmpra9Xc3KzIyEjV1dW1HTs8fvy4Ro8erdraWpdTAgAQHK666ioVFRUpIyOj3dqFCxcUExND4wcAegEaP13rsTN+9u/ff8XclNWrVysvL095eXl69tlnHa//6quvasSIESosLFRzc7O+8pWvyFqrwsJCjRw5Uq+99prjGeA9n376qcLDw9WnTx/FxMRcMWtq4MCBqq+vdzEdAADBJSMjQzt37lRoaGi7R1hYGEe9AABBzfMzfqZOnarS0lLFxcVp6dKlWr9+ve6++25J0i9/+UsdPnxYS5Yscaz+vffeq9WrV2vatGnt1oqLi7VgwQJ98MEHjtWHNyUnJ+vIkSNKSkpSWVnZFWuVlZVKSEhwKRkAAMHn6aef7nT2Yp8+fdTc3BzgRAAAeIfnj3pFRUWpoaFBknT11Vfrb3/7m4YNGyZJqqmp0fjx43Xw4EFHskpSv379dPLkSfXp06fdmt/vV//+/XX+/HnH6sObtm7dqtTU1A4bPMXFxTp8+LAWLlzoQjIAAAAA6J046tW1HnvUa9CgQdq3b58kqbGxUfHx8W1rcXFxOnXqlKP1r7/+ej300EM6d+7cFc+fO3dODz/8sK6//npH68ObMjMzO93Vk5ubS9MHAIAAO3r06BW/DJSWlmrFihXatWuXi6kAAHCf5xs/ixYt0u23367XXntNP/3pTzVjxgxt27ZN27Zt04wZM5STk+No/TVr1mj79u0aOHCgUlNTdcMNNygtLU0JCQnatm2b1q5d62h9eJPbs6cAAMAlJSUlGjlypG688Ubl5ORoxYoVeuqpp1RaWqpx48appKTE7YgAALjG80e9JOkPf/iD8vPzVVNT0/ZcRESEpk+froKCAkVHR3d3zHYqKytVUVGhhoYGRUVFKTU1VSNHjnS8LrwpNTVVpaWlGj58eLvZU6tWrdLMmTMdnT0FAAAu+eY3v6mnn35akjRu3Dht3rxZkydPliQVFhaqoKBAO3bscDMiAKAbcNSraz3+du7WWtXU1KimpkY+n0/XXHONfD6fAwk/X319vay1iomJcaU+3Of27CkAAHBJXFxc205cn8+nhoYGhYaGSmq5E+egQYN08uRJNyMCALoBjZ+u9dgZPxcZY5ScnKxvf/vbGjNmTMCaPgUFBW0zhg4dOqQJEyYoNjZW/fv316RJk3T48OGA5IC3uD17CgAAXBIREdF2566srKy2po/U0vhpampyKxoAAK7rMY2fjjQ2Nuqmm25ytMby5cs1ZMgQSdLChQt13XXX6cSJEzp+/LjS09O1YMECR+vDm9yePQUAAC5JS0vT3r17JUkbNmy4Ym3r1q269tpr3YgFAIAn9JijXh3x+/2KjIx09H9xoqOjderUKYWFhSkxMVGHDh1SRESEpJbGU1JSkurq6hyrD+/ywuwpAADQtX379slaq6997WtuRwEAfEkc9epaZ0e9wtwI889ISUnpdO3ill4npaen68UXX9T06dM1YsQIlZeXKz09XZL0wQcfuDZnCO778Y9/rPnz53tm9hQAAGjvmmuucTsCAACu8nzj5+TJk/r1r3+t4cOHt1v75JNPdOuttzpa/7e//a2ys7O1efNmjR07VllZWcrNzZUxRsXFxVq+fLmj9eFtF2dPJScnS5Kqq6uVnJwsY4zLyQAACC5r167VM888o/Lycp0/f15DhgzRt771LT388MPs9gEABDXPN37S09Pl8/l08803t1vz+/1y+qhaenq69u7dqyeeeEJvvfWWEhIS9N577+kb3/iGXnzxRX3nO99xtD56lnHjxmn37t1KSEhwOwoAAEHj0Ucf1XPPPaf58+fre9/7ntasWaM77rhDTU1NuvHGG7Vu3TplZWW5HRMAAFd4fsbP1q1b1a9fP40dO7bdmrVWr7/+uiZNmtSdEYHPNXTo0A6fr62tVVJSkkJCQlRdXR3gVAAABKekpCTt2LFDw4YNkyTt379fubm52rNnjzZv3qwlS5bo/fffdzklAODLYsZP13rsjJ/MzMxO14wxAW/67Nu3Ty+//LIkafLkyRo1alRA68MbIiMjlZSUpJ///Ofq27evpJZGZF5engoKCq64vTsAAHCWtVZJSUlt10lJSTpx4oSkltdrt912m1vRAABwXY++nbskx3dVZGZmaufOnZKkoqIiZWRkaMuWLdqyZYvGjh2rkpISR+vDm/bs2aPvfve7uu+++3TmzBlNmjRJmZmZioiI0Pjx49mFBgBAAGVnZ2vWrFl6//339e6772rOnDltYwJOnz6t2NhYlxMCAOCeHt348fv9HQ597k67d+/WmDFjJLWcH9+4caNKSkpUUlKiTZs26cEHH3S0PrwpPDxcS5YsUWlpqVatWqUpU6aoqqqKoc4AALhgxYoVioyM1OTJk5WdnS2fz6eCggJJLY2fJ5980uWEAAC4x/NHvV5//fVO1/x+v+P1w8LCdOrUKcXHx6u2tlY33HBD29q4ceN08OBBxzPAu4YPH64NGzaouLhY2dnZqqurczsSAABBJzY2VmvWrOlwLSUlRSkpKQFOBACAd3i+8ZOZmanBgwcrJMSdzUm33XabFi5cqJUrV+rOO+/UL37xCz300EOy1urxxx/X17/+dVdywVtyc3M1efJkVVZWasCAAW7HAQAgKH344YeqqKhQfX29oqOjlZqayjxGAEDQ83zjZ9iwYXr++eev2Glz0YULF9SvX7uB1d3qN7/5je666y4NGTJEQ4cOVXl5uZYtWyap5c5Of/7znx2tj56jublZKSkpCg0NdTsKAABBpbq6Wrfffrt2796tq6++WrGxsTp79qyqqqo0evRovfDCC53ekRMAgN7O8zN+urpdW0hIiOM/xH0+nwoLC/X222/r/vvv17Jly/T4449r06ZN2rt3r6699lpH68ObLjb/JKmurk633HKLYmJiFBcXp6ysLB07dszFdAAABJe5c+dq4sSJOnHihPbs2aNt27aprKxMx44d08SJEzVnzhy3IwIA4BpjrQ1owYyMDNtZI6cjjY2NklqG6bqlurpa7733XofbhdevX6877rjDpWRwS0xMjM6ePStJmjdvnurr6/W73/1OkrRo0SL17dtXzz77rIsJAQAIHlFRUTp58qQiIiLarfn9fsXHx+vcuXMuJAMAdKeuNoZAMsact9a2Oxbl+R0/4eHhnTZ9mpqalJ+f72j9v/zlL0pLS9PSpUs1ZswYLViwQE1NTW3rd999t6P14U2XN0xfeeUV/f73v1diYqISExP11FNPacuWLS6mAwAguCQnJ6u0tLTDtU2bNnHMCwAQ1Dy/46crfr9fkZGRVzRiutt1112n/Px85eTk6OjRo5o5c6b69OmjoqIiRUREKDo6WvX19Y7VhzfFxMTozJkzam5u1le/+lUdPHiwbQC5tVYxMTF8XwAAECCvvvqq8vLylJaWptGjR7fN+Nm1a5cqKir00ksv6aabbnI7JgDgS2LHT9c62/Hj+eHO8+bN63Tt008/dbz+/v37lZOTI0lKTEzU5s2bNXPmTE2ZMkUlJSWO14c3NTQ0KCwsTNZaGWO0a9cupaenS5IqKyuVkJDgckIAAILHzTffrKqqKhUVFamiokLHjh1TVFSUZs+erdzcXA0cONDtiAAAuMbzjZ9169Zp/vz5io+Pb7fm5E6fi/r3769Dhw4pOTlZkhQWFqb169dr/vz5ysrKCkgGeM+BAweuuL78BeXp06e1fPnyQEcCACCoDRgwQHfddZeklp/TGzdulN/v1+nTp2n8AACCWsCPehljmiX945/4kL6SGiV11mGJlHT+y+bqQoQk25qho7Uwh+ujZwpXx98zAACg+/l06fVliFpeP1587Rgq6YKkZhdyAQC6l9O///d0Pmttu1nOAW/8/LOMMT+RdNha+z8drIVKesha+6iD9SMkhVlrO/zmMsYMtdZWO1UfPY8xpo+k89baULezAAAQDIwx9dba6Na335C0ylr7x9brH0n6ibX2BjczAgDgFs83fgAvMsY808VymKQf0fgBACAwjDFnrbUxrW8fk3SVtbax9TpU0nFrbfu5AQAABAHPz/gBPGqGpNWSTnawRsMHAIDACjfGzJVk1HJEP0KXjlyHiZ/NAIAgRuMH+GL2SPqrtbbdrd2MMX0lPRD4SAAABK0dkma1vr1X0r9Ieqf1epKkfW6EAgDAC2j8AF/Ms2oZHtmRRkmOzZ0CAABXstZmdrG8Q9KUAEUBAMBzmPEDAAAAAADQS3W2YwEAAAAAAAA9HI0fAAAAAACAXorGDwAA6BbGmI+NMY8YYxpaHxeMMU2XXVe0vp81xpy77PkGY8x/ta4tbV3/98987v9ofX7pZc89aIw50PrxNcaYPwX0CwYAAOgBaPwAAIDu9Ka1NspaGyXp3yT938Vra23qZe83+rLno6y1v7ps7UNJsz/zeWe1Pi9JMsbMlnSnpKzWWhmSXnXkKwIAAOjBaPwAAACveUdSpDEmVZJa//Tp0u25JWmspL9aa6skyVp7xFq7MuBJAQAAPI7GDwAA8KLn1LLLR2rZ/fPHz6y/JWmWMeZ+Y0yGMSY0oOkAAAB6CBo/AADADe8bY05f9pj8mfVCSXcYOOOOXQAAATtJREFUY8IlTW+9bmOtLZR0r6TJkv5X0jFjzAOBCA4AANCThLkdAAAABKV0a+3+zhattdXGmP2SlkuqtNYeMsZ89n2el/R8a3PoX1vf3mmt/auTwQEAAHoSdvwAAACv+qOk/1T7Y15XsNY2Wmv/W1KZpLRABAMAAOgp2PEDAAC86k+SaiRt/+yCMWaOpOOSXpd0Ti1HvlIl7QhgPgAAAM9jxw8AAHDDbmNMw2WPJz/7Dtbaf1hrX7HW/qODjz8r6UFJ1ZJOS/qVpHustducjQ0AANCzGGut2xkAAAAAAADgAHb8AAAAAAAA9FI0fgAAAAAAAHopGj8AAAAAAAC9FI0fAAAAAACAXorGDwAAAAAAQC9F4wcAAAAAAKCXovEDAAAAAADQS9H4AQAAAAAA6KVo/AAAAAAAAPRS/w+8S1AmPf8nBQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1440x720 with 3 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "###Optional\n",
    "#make a supervenn to plot the overlaps (takes some time)\n",
    "try:\n",
    "    from supervenn import supervenn\n",
    "except ModuleNotFoundError:\n",
    "    !pip install supervenn #does not work for some reason it looks at the base env for this package (print(sys.executable) )\n",
    "    from supervenn import supervenn\n",
    "\n",
    "labels = list(proteins.keys())\n",
    "sets = [proteins[x] for x in proteins.keys()]\n",
    "\n",
    "plt.figure(figsize=(20, 10))\n",
    "supervenn(sets, labels, min_width_for_annotation=10000, rotate_col_annotations=True, \n",
    "          side_plots='right', chunks_ordering='occurrence', alternating_background= False)\n",
    "plt.grid(False)\n",
    "plt.savefig('baroverlap.svg')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52842789",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "source": [
    "#### Can you also show peptides from the six-frame translation? can you speculate how they would look like?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2be6514f",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "# Part 3: DNA Variants to Proteins\n",
    "\n",
    "\n",
    "In this part we will use the `vcf-to-proteindb` command to translate mutations in VCF format to proteins. The output of genome/exome sequencing is usually a VCF file that includes the list of identified mutations.\n",
    "\n",
    "\n",
    "## Download relevant files\n",
    "\n",
    "Lets download a `VCF` file from `ENSEMBL` containing all `SNP`s from homo sapiens, to minimize computational time we only download SNPs identified in `chrY`. \n",
    "\n",
    "The idea is to learn how to convert a `VCF` file and you can replace this `VCF` with any standard `VCF` file that is generated from WGS/WES. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3cf26272",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2022-06-09 10:49:30--  http://ftp.ensembl.org/pub/release-106/variation/vcf/homo_sapiens/homo_sapiens_incl_consequences-chrY.vcf.gz\n",
      "Resolving ftp.ensembl.org (ftp.ensembl.org)... 193.62.193.139\n",
      "Connecting to ftp.ensembl.org (ftp.ensembl.org)|193.62.193.139|:80... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 31480732 (30M) [application/x-gzip]\n",
      "Saving to: homo_sapiens_incl_consequences-chrY.vcf.gz\n",
      "\n",
      "homo_sapiens_incl_c 100%[===================>]  30,02M  3,95MB/s    in 8,2s    \n",
      "\n",
      "2022-06-09 10:49:38 (3,65 MB/s) - homo_sapiens_incl_consequences-chrY.vcf.gz saved [31480732/31480732]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget http://ftp.ensembl.org/pub/release-106/variation/vcf/homo_sapiens/homo_sapiens_incl_consequences-chrY.vcf.gz\n",
    "!gunzip homo_sapiens_incl_consequences-chrY.vcf.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d25dfd4",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "source": [
    "Also, we need to download transcript structures and sequences for homo sapiens, run the following to download `GTF` and `cdna` files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "2b3e34f4",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n",
      "DEBUG:pypgatk.toolbox.general:Files downloaded -- ensembl_files/Homo_sapiens.GRCh38.106.gtf,ensembl_files/Homo_sapiens.GRCh38.cds.all.fa\n",
      "INFO:Main Pipeline Ensembl Downloader:Pipeline Finish !!!\n"
     ]
    }
   ],
   "source": [
    "!pypgatk_cli.py ensembl-downloader -c config/ensembl_downloader_config.yaml -t 9606 -o ensembl_files --skip_vcf --skip_protein --skip_ncrna --skip_dna --skip_cdna\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08d760e8",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "#### Check the flags that can be used for a VCF file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "4b78830f",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n",
      "Usage: pypgatk_cli.py vcf-to-proteindb [OPTIONS]\n",
      "\n",
      "Options:\n",
      "  -c, --config_file TEXT          Configuration to perform conversion between\n",
      "                                  ENSEMBL Files\n",
      "  -f, --input_fasta TEXT          Path to the transcript sequence\n",
      "  -v, --vcf TEXT                  Path to the VCF file\n",
      "  -g, --gene_annotations_gtf TEXT\n",
      "                                  Path to the gene annotations file\n",
      "  -t, --translation_table INTEGER\n",
      "                                  Translation table (Default 1)\n",
      "  -m, --mito_translation_table INTEGER\n",
      "                                  Mito_trans_table (default 2)\n",
      "  -p, --var_prefix TEXT           String to add before the variant peptides\n",
      "  --report_ref_seq                In addition to var peps, also report all ref\n",
      "                                  peps\n",
      "  -o, --output_proteindb TEXT     Output file name, exits if already exists\n",
      "  --annotation_field_name TEXT    Annotation field name found in the INFO\n",
      "                                  column,\n",
      "                                  e.g CSQ or vep; if empty it will\n",
      "                                  identify overlapping transcripts\n",
      "                                  from the\n",
      "                                  given GTF file and no aa consequence will be\n",
      "                                  considered\n",
      "  --af_field TEXT                 field name in the VCF INFO column to use for\n",
      "                                  filtering on AF, (Default None)\n",
      "  --af_threshold FLOAT            Minium AF threshold for considering common\n",
      "                                  variants\n",
      "  --transcript_str TEXT           String that is used for transcript ID in the\n",
      "                                  VCF header INFO field\n",
      "  --biotype_str TEXT              String that is used for biotype in the VCF\n",
      "                                  header INFO field\n",
      "  --exclude_biotypes TEXT         Excluded Biotypes  [default: ]\n",
      "  --include_biotypes TEXT         included_biotypes, default all\n",
      "  --consequence_str TEXT          String that is used for consequence in the\n",
      "                                  VCF header INFO field\n",
      "  --exclude_consequences TEXT     Excluded Consequences  [default:\n",
      "                                  downstream_gene_variant,\n",
      "                                  upstream_gene_variant, intergenic_variant,\n",
      "                                  intron_variant, synonymous_variant]\n",
      "  -s, --skip_including_all_cds    by default any transcript that has a defined\n",
      "                                  CDS will be used, this option disables this\n",
      "                                  features instead\n",
      "  --include_consequences TEXT     included_consequences, default all\n",
      "  --ignore_filters                enabling this option causes or variants to\n",
      "                                  be parsed. By default only variants that\n",
      "                                  have not failed any filters will be\n",
      "                                  processed (FILTER column is PASS, None, .)\n",
      "                                  or if the filters are subset of the accepted\n",
      "                                  filters. (default is False)\n",
      "  --accepted_filters TEXT         Accepted filters for variant parsing\n",
      "  -h, --help                      Show this message and exit.\n"
     ]
    }
   ],
   "source": [
    "!pypgatk_cli.py vcf-to-proteindb -h"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36ff542f",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "#### Convert the variants to protein sequences\n",
    "\n",
    "\n",
    "The first time you run the command, it takes a bit long because it converts the `GTF` file to an `sqlite` database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "3863cf00",
   "metadata": {
    "scrolled": true,
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n",
      "2022-06-09 10:59:53,905 - INFO - Committing changes: 3279000 features\n",
      "2022-06-09 10:59:55,943 - INFO - Populating features table and first-order relations: 3279404 features\n",
      "2022-06-09 10:59:55,946 - INFO - Creating relations(parent) index\n",
      "2022-06-09 11:00:07,480 - INFO - Creating relations(child) index\n",
      "2022-06-09 11:00:18,876 - INFO - Creating features(featuretype) index\n",
      "2022-06-09 11:00:36,776 - INFO - Creating features (seqid, start, end) index\n",
      "2022-06-09 11:00:57,518 - INFO - Creating features (seqid, start, end, strand) index\n",
      "2022-06-09 11:01:18,656 - INFO - Running ANALYZE features\n",
      "/Users/husen/opt/anaconda3/envs/pygatk/lib/python3.8/site-packages/Bio/Seq.py:2606: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\"Partial codon, len(sequence) not a multiple of three. \"\n",
      "Translation summary:\n",
      " # variants with invalid record:3430\n",
      "# variants not passing Filter:0\n",
      "# variants not passing AF threshold:0\n",
      "# feature IDs from VCF that are not found in the given FASTA file:1153573\n",
      "# variants successfully translated:4485\n",
      "Generated dict_items([('var', 28353)]) variant proteins from 4479 variants\n"
     ]
    }
   ],
   "source": [
    "!pypgatk_cli.py vcf-to-proteindb -c config/ensembl_config.yaml --input_fasta ensembl_files/Homo_sapiens.GRCh38.cds.all.fa --vcf homo_sapiens_incl_consequences-chrY.vcf --gene_annotations_gtf ensembl_files/Homo_sapiens.GRCh38.106.gtf --output_proteindb chrY_protein_variants.fa\n",
    "\n",
    "print('Generated {} variant proteins from {} variants'.format(\n",
    "    Counter([x.split('_')[0] for x in list(SeqIO.index('chrY_protein_variants.fa', \"fasta\").keys())]).items(),\n",
    "    len(set(['_'.join(x.split('_')[0:3]) for x in SeqIO.index('chrY_protein_variants.fa', \"fasta\").keys()]))))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1abb6452",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "source": [
    "#### Filter variants by allele frequency and consequence \n",
    "\n",
    "Depending on the VCF file, you may also need to specify which field name indicates allele frequency using the `--af_field` parameter.\n",
    "\n",
    "Below, we only variants that are common in the population (MAF>=0.01)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "b42b1707",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n",
      "Database already exists: table features already exists ensembl_files/Homo_sapiens.GRCh38.106.db\n",
      "/Users/husen/opt/anaconda3/envs/pygatk/lib/python3.8/site-packages/Bio/Seq.py:2606: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\"Partial codon, len(sequence) not a multiple of three. \"\n",
      "Translation summary:\n",
      " # variants with invalid record:1503251\n",
      "# variants not passing Filter:0\n",
      "# variants not passing AF threshold:132214\n",
      "# feature IDs from VCF that are not found in the given FASTA file:28860\n",
      "# variants successfully translated:15\n",
      "Generated dict_items([('var', 66)]) variant proteins from 17 variants\n"
     ]
    }
   ],
   "source": [
    "!pypgatk_cli.py vcf-to-proteindb -c config/ensembl_config.yaml --input_fasta ensembl_files/Homo_sapiens.GRCh38.cds.all.fa --vcf homo_sapiens_incl_consequences-chrY.vcf --gene_annotations_gtf ensembl_files/Homo_sapiens.GRCh38.106.gtf --output_proteindb chrY_protein_variants_VAF05.fa --af_field 'MAF' --af_threshold 0.01\n",
    "\n",
    "print('Generated {} variant proteins from {} variants'.format(\n",
    "    Counter([x.split('_')[0] for x in list(SeqIO.index('chrY_protein_variants_VAF05.fa', \"fasta\").keys())]).items(),\n",
    "    len(set(['_'.join(x.split('_')[0:3]) for x in SeqIO.index('chrY_protein_variants_VAF05.fa', \"fasta\").keys()]))))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5893e2a3",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "source": [
    "Filter by consequence, only translate frameshift variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "ddc9a335",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n",
      "Database already exists: table features already exists ensembl_files/Homo_sapiens.GRCh38.106.db\n",
      "/Users/husen/opt/anaconda3/envs/pygatk/lib/python3.8/site-packages/Bio/Seq.py:2606: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\"Partial codon, len(sequence) not a multiple of three. \"\n",
      "Translation summary:\n",
      " # variants with invalid record:3430\n",
      "# variants not passing Filter:0\n",
      "# variants not passing AF threshold:0\n",
      "# feature IDs from VCF that are not found in the given FASTA file:1153573\n",
      "# variants successfully translated:2698\n",
      "Generated dict_items([('var', 21418)]) variant proteins from 2736 variants\n"
     ]
    }
   ],
   "source": [
    "!pypgatk_cli.py vcf-to-proteindb -c config/ensembl_config.yaml --input_fasta ensembl_files/Homo_sapiens.GRCh38.cds.all.fa --vcf homo_sapiens_incl_consequences-chrY.vcf --gene_annotations_gtf ensembl_files/Homo_sapiens.GRCh38.106.gtf --output_proteindb chrY_protein_variants_framshift.fa --include_consequences 'frameshift_variant,missense_variant'\n",
    "\n",
    "print('Generated {} variant proteins from {} variants'.format(\n",
    "    Counter([x.split('_')[0] for x in list(SeqIO.index('chrY_protein_variants_framshift.fa', \"fasta\").keys())]).items(),\n",
    "    len(set(['_'.join(x.split('_')[0:3]) for x in SeqIO.index('chrY_protein_variants_framshift.fa', \"fasta\").keys()]))))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fb8a422",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "source": [
    "### * Note on translating patient sequencing data\n",
    "The same command as above can be applied to generate a protein database per patient.\n",
    "\n",
    "Provide path of the VCF file that you obtain from `WGS`/`WES`. \n",
    "\n",
    "If you also have `RNA-seq` from the same patient, you can convert the `GTF` file to `fasta` using the gffread command shown in the second part of this notebook. This enables translation of expressed transcripts only.\n",
    "\n",
    "For cases where no `RNA-seq` is perfromed, you can still give the refernce GTF file and the canonical transcripts will be used to generate mutated/variant proteins"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61404e7e",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "# Part 4a: Mutations to Proteins (cBioPortal)\n",
    "\n",
    "\n",
    "In this part we will use the `cbioportal-downloader` and `cbioportal-to-proteindb` command to translate mutations from cancer studies in cBioPortal (https://www.cbioportal.org/).\n",
    "\n",
    "\n",
    "First we need to specify what study to download from cBioPortal (see -l flag): \n",
    "\n",
    "\n",
    "\n",
    "Check the flags below to know what flags to use."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "3c3eddfc",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n",
      "Usage: pypgatk_cli.py cbioportal-downloader [OPTIONS]\n",
      "\n",
      "Options:\n",
      "  -c, --config_file TEXT       Configuration file for the ensembl data\n",
      "                               downloader pipeline\n",
      "  -o, --output_directory TEXT  Output directory for the peptide databases\n",
      "  -l, --list_studies           Print the list of all the studies in cBioPortal\n",
      "                               (https://www.cbioportal.org)\n",
      "  -d, --download_study TEXT    Download a specific Study from cBioPortal --\n",
      "                               (all to download all studies)\n",
      "  -th, --multithreading        Enable multithreading to download multiple\n",
      "                               files ad the same time\n",
      "  -h, --help                   Show this message and exit.\n"
     ]
    }
   ],
   "source": [
    "!pypgatk_cli.py cbioportal-downloader -h"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "572f26ec",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "source": [
    "#### Download data for study ID blca_mskcc_solit_2014\n",
    "\n",
    "In addition to the files from cBioPortal we also need to download coding sequences from GRCh37, the matching genome assemblies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "c700b496",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n",
      "DEBUG:pypgatk.toolbox.general:The following study 'blca_mskcc_solit_2014' has been downloaded. \n",
      "x blca_mskcc_solit_2014/\n",
      "x blca_mskcc_solit_2014/meta_mutations.txt\n",
      "x blca_mskcc_solit_2014/data_log2_cna.txt\n",
      "x blca_mskcc_solit_2014/Readme.txt\n",
      "x blca_mskcc_solit_2014/meta_clinical_sample.txt\n",
      "x blca_mskcc_solit_2014/data_cna.txt\n",
      "x blca_mskcc_solit_2014/case_lists/\n",
      "x blca_mskcc_solit_2014/case_lists/cases_sequenced.txt\n",
      "x blca_mskcc_solit_2014/case_lists/cases_all.txt\n",
      "x blca_mskcc_solit_2014/case_lists/cases_cnaseq.txt\n",
      "x blca_mskcc_solit_2014/case_lists/cases_log2CNA.txt\n",
      "x blca_mskcc_solit_2014/case_lists/cases_cna.txt\n",
      "x blca_mskcc_solit_2014/data_clinical_patient.txt\n",
      "x blca_mskcc_solit_2014/data_clinical_sample.txt\n",
      "x blca_mskcc_solit_2014/meta_cna.txt\n",
      "x blca_mskcc_solit_2014/data_cna_hg19.seg\n",
      "x blca_mskcc_solit_2014/data_gene_panel_matrix.txt\n",
      "x blca_mskcc_solit_2014/meta_clinical_patient.txt\n",
      "x blca_mskcc_solit_2014/meta_study.txt\n",
      "x blca_mskcc_solit_2014/LICENSE\n",
      "x blca_mskcc_solit_2014/data_mutations.txt\n",
      "x blca_mskcc_solit_2014/meta_gene_panel_matrix.txt\n",
      "x blca_mskcc_solit_2014/meta_cna_hg19_seg.txt\n",
      "x blca_mskcc_solit_2014/meta_log2_cna.txt\n",
      "--2022-06-09 12:15:50--  ftp://ftp.ensembl.org/pub/release-75/fasta/homo_sapiens/cds/Homo_sapiens.GRCh37.75.cds.all.fa.gz\n",
      "           => Homo_sapiens.GRCh37.75.cds.all.fa.gz\n",
      "Resolving ftp.ensembl.org (ftp.ensembl.org)... 193.62.193.139\n",
      "Connecting to ftp.ensembl.org (ftp.ensembl.org)|193.62.193.139|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /pub/release-75/fasta/homo_sapiens/cds ... done.\n",
      "==> SIZE Homo_sapiens.GRCh37.75.cds.all.fa.gz ... 19834207\n",
      "==> PASV ... done.    ==> RETR Homo_sapiens.GRCh37.75.cds.all.fa.gz ... done.\n",
      "Length: 19834207 (19M) (unauthoritative)\n",
      "\n",
      "Homo_sapiens.GRCh37 100%[===================>]  18,92M  10,3MB/s    in 1,8s    \n",
      "\n",
      "2022-06-09 12:15:52 (10,3 MB/s) - Homo_sapiens.GRCh37.75.cds.all.fa.gz saved [19834207]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "study_id = 'blca_mskcc_solit_2014'\n",
    "!pypgatk_cli.py cbioportal-downloader -c config/cbioportal_config.yaml -d $study_id -o ./\n",
    "!tar -xzvf $study_id\".tar\"\n",
    "\n",
    "#Download linking CDSs (the mutations are aligned to hg37)\n",
    "!wget ftp://ftp.ensembl.org/pub/release-75/fasta/homo_sapiens/cds/Homo_sapiens.GRCh37.75.cds.all.fa.gz\n",
    "!gunzip *.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1a42ac4",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "source": [
    "The files are downloaded into a directory with the same name as the study ID. It contains several files but we only need the following two files:\n",
    "- `data_clinical_sample.txt`: \n",
    "\n",
    "    This file contains metada or clinical information about each of the samples in the study\n",
    "    \n",
    "    \n",
    "- `data_mutations.txt` (it might have slightly different names depending on the cBioPortal version): \n",
    "\n",
    "    This file contains the list of mutations identified in this study\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b31ddfcb",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "source": [
    "#### Translate the downloaded set of mutations\n",
    "\n",
    "First check what options you have!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "fd4d03f8",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n",
      "Usage: pypgatk_cli.py cbioportal-to-proteindb [OPTIONS]\n",
      "\n",
      "Options:\n",
      "  -c, --config_file TEXT          Configuration for cbioportal to proteindb\n",
      "                                  tool\n",
      "  -in, --input_mutation TEXT      Cbioportal mutation file\n",
      "  -fa, --input_cds TEXT           CDS genes from ENSEMBL database\n",
      "  -out, --output_db TEXT          Protein database including all the mutations\n",
      "  -f, --filter_column TEXT        Column in the VCF file to be used for\n",
      "                                  filtering or splitting mutations\n",
      "  -a, --accepted_values TEXT      Limit mutations to values (tissue type,\n",
      "                                  sample name, etc) considered for generating\n",
      "                                  proteinDBs, by default mutations from all\n",
      "                                  records are considered\n",
      "  -s, --split_by_filter_column    Use this flag to generate a proteinDB per\n",
      "                                  group as specified in the filter_column,\n",
      "                                  default is False\n",
      "  -cl, --clinical_sample_file TEXT\n",
      "                                  File to get tissue type from for the samples\n",
      "                                  in input_mutation file\n",
      "  -h, --help                      Show this message and exit.\n"
     ]
    }
   ],
   "source": [
    "!pypgatk_cli.py cbioportal-to-proteindb -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "3581314b",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n",
      "/Users/husen/opt/anaconda3/envs/pygatk/lib/python3.8/site-packages/Bio/Seq.py:2606: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\"Partial codon, len(sequence) not a multiple of three. \"\n",
      "Generated Counter({'cbiomut': 1125}) variant proteins from 1057 variants\n"
     ]
    }
   ],
   "source": [
    "\n",
    "!pypgatk_cli.py cbioportal-to-proteindb --config_file config/cbioportal_config.yaml --input_cds Homo_sapiens.GRCh37.75.cds.all.fa --input_mutation $study_id/data_mutations.txt --clinical_sample_file blca_mskcc_solit_2014/data_clinical_sample.txt --output_db variants_proteindb.fa\n",
    "\n",
    "print('Generated {} variant proteins from {} variants'.format(\n",
    "    Counter([str(x.id).split(':')[0] for x in SeqIO.parse('variants_proteindb.fa', \"fasta\")]),\n",
    "    len(set([x.id for x in SeqIO.parse('variants_proteindb.fa', \"fasta\")]))\n",
    "    ))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "612fca86",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "source": [
    "Optionally, you can filter the samples based on any of the columns in the data_clinical_sample.txt. First, let's see what fields (clinical information) is available for the samples in this study:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "1c575e13",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['#Patient Identifier',\n",
       " 'Sample Identifier',\n",
       " 'Sample Type',\n",
       " 'Concomitant Carcinoma in situ',\n",
       " 'Lymph Node Status',\n",
       " 'Tissue Sequenced',\n",
       " 'Oncotree Code',\n",
       " 'Cancer Type',\n",
       " 'Cancer Type Detailed',\n",
       " 'Somatic Status',\n",
       " 'TMB (nonsynonymous)']"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[x.strip() for x in open('{}/data_clinical_sample.txt'.format(study_id)).readline().split('\\t')]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9284606f",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "source": [
    "Let's only use mutations from samples that have their `Lymph Node Status` as `Positive`. Cancer Type can also be used to filter mutations for a specific cancer type."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "5a9a7c87",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n",
      "sample_groups_dict blca_mskcc_solit_2014/data_clinical_sample.txt LYMPH_NODE_STATUS\n",
      "No clinical info was found for sample Tumor_Sample_Barcode. Skipping (line 0): Hugo_Symbol\tEntrez_Gene_Id\tCenter\tNCBI_Build\tChromosome\tStart_Position\tEnd_Position\tStrand\tConsequence\tVariant_Classification\tVariant_Type\tReference_Allele\tTumor_Seq_Allele1\tTumor_Seq_Allele2\tdbSNP_RS\tdbSNP_Val_Status\tTumor_Sample_Barcode\tMatched_Norm_Sample_Barcode\tMatch_Norm_Seq_Allele1\tMatch_Norm_Seq_Allele2\tTumor_Validation_Allele1\tTumor_Validation_Allele2\tMatch_Norm_Validation_Allele1\tMatch_Norm_Validation_Allele2\tVerification_Status\tValidation_Status\tMutation_Status\tSequencing_Phase\tSequence_Source\tValidation_Method\tScore\tBAM_File\tSequencer\tt_ref_count\tt_alt_count\tn_ref_count\tn_alt_count\tHGVSc\tHGVSp\tHGVSp_Short\tTranscript_ID\tRefSeq\tProtein_position\tCodons\tHotspot\n",
      "\n",
      "/Users/husen/opt/anaconda3/envs/pygatk/lib/python3.8/site-packages/Bio/Seq.py:2606: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\"Partial codon, len(sequence) not a multiple of three. \"\n",
      "Generated Counter({'cbiomut': 363}) variant proteins from 353 variants\n"
     ]
    }
   ],
   "source": [
    "\n",
    "!pypgatk_cli.py cbioportal-to-proteindb --config_file config/cbioportal_config.yaml --input_cds Homo_sapiens.GRCh37.75.cds.all.fa --input_mutation $study_id/data_mutations.txt --clinical_sample_file blca_mskcc_solit_2014/data_clinical_sample.txt --output_db variants_proteindb_filtered.fa --filter_column 'LYMPH_NODE_STATUS' --accepted_values 'Positive'\n",
    "\n",
    "print('Generated {} variant proteins from {} variants'.format(\n",
    "    Counter([str(x.id).split(':')[0] for x in SeqIO.parse('variants_proteindb_filtered.fa', \"fasta\")]),\n",
    "    len(set([x.id for x in SeqIO.parse('variants_proteindb_filtered.fa', \"fasta\")]))\n",
    "    ))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f2815bf",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "# Part 4b: Mutations to Proteins (COSMIC)\n",
    "\n",
    "\n",
    "In this part we will use the `cosmic-downloader` and `cosmic-to-proteindb` command to translate cancer mutations from COSMIC. The downloader downloads the following files:\n",
    "\n",
    "- `CosmicMutantExport.tsv`  : `tsv` file containing somatic mutations from human samples\n",
    "-  `CosmicCLP_MutantExport.tsv` : `tsv` file containing somatic mutations from cancer cell lines\n",
    "- `All_COSMIC_Genes.fasta`: Transcripts `fasta` file, it is used to extract sequences of altered transcripts.\n",
    "\n",
    "However, since COSMIC requires registeration and the file sizes are large (15G), the download takes some time therefore I have pre-downloaded the files and extracted 50k mutations from it as a sample to make the processes run faster. Run the following code to download a sample directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "599af02e",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2022-06-09 12:16:04--  https://github.com/husensofteng/ProteoWorks/raw/master/workshop_pgdb/cosmic_files.tar.gz\n",
      "Resolving github.com (github.com)... 140.82.121.3\n",
      "Connecting to github.com (github.com)|140.82.121.3|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 Found\n",
      "Location: https://raw.githubusercontent.com/husensofteng/ProteoWorks/master/workshop_pgdb/cosmic_files.tar.gz [following]\n",
      "--2022-06-09 12:16:05--  https://raw.githubusercontent.com/husensofteng/ProteoWorks/master/workshop_pgdb/cosmic_files.tar.gz\n",
      "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.110.133, 185.199.111.133, ...\n",
      "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 38247627 (36M) [application/octet-stream]\n",
      "Saving to: cosmic_files.tar.gz\n",
      "\n",
      "cosmic_files.tar.gz 100%[===================>]  36,48M  11,6MB/s    in 3,2s    \n",
      "\n",
      "2022-06-09 12:16:10 (11,6 MB/s) - cosmic_files.tar.gz saved [38247627/38247627]\n",
      "\n",
      "x cosmic_files/\n",
      "x cosmic_files/All_CellLines_Genes.fasta\n",
      "x cosmic_files/CosmicCLP_MutantExport.tsv\n",
      "x cosmic_files/CosmicMutantExport.tsv\n",
      "x cosmic_files/All_COSMIC_Genes.fasta\n"
     ]
    }
   ],
   "source": [
    "if not os.path.isfile('cosmic_files.tar.gz'):\n",
    "    !wget https://github.com/husensofteng/ProteoWorks/raw/master/workshop_pgdb/cosmic_files.tar.gz\n",
    "    !tar -xzvf cosmic_files.tar.gz\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "862c3b8e",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "source": [
    "### For actual analysis:\n",
    "- remove the sample `cosmic_files` directory\n",
    "- register an account on: https://cancer.sanger.ac.uk/\n",
    "- replace `yourUserName` and `yourPassword` in the follwing command with your actual credentials from your account in COSMIC. \n",
    "- run the following command to download the mutation data files from COSMIC and ponder on some simple childhood memories until the files are downloaded :)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "a6a41a60",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [],
   "source": [
    "if not os.path.isdir('cosmic_files'):\n",
    "    !pypgatk_cli.py cosmic-downloader -u yourUserName -p yourPassword -c config/cosmic_config.yaml -o cosmic_files/\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "499e7dca",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n",
      "Usage: pypgatk_cli.py cosmic-to-proteindb [OPTIONS]\n",
      "\n",
      "Options:\n",
      "  -c, --config_file TEXT        Configuration file for the cosmic data\n",
      "                                pipelines\n",
      "  -in, --input_mutation TEXT    Cosmic Mutation data file\n",
      "  -fa, --input_genes TEXT       All Cosmic genes\n",
      "  -out, --output_db TEXT        Protein database including all the mutations\n",
      "  -f, --filter_column TEXT      Column in the VCF file to be used for\n",
      "                                filtering or splitting mutations\n",
      "  -a, --accepted_values TEXT    Limit mutations to values (tissue type, sample\n",
      "                                name, etc) considered for generating\n",
      "                                proteinDBs, by default mutations from all\n",
      "                                records are considered\n",
      "  -s, --split_by_filter_column  Use this flag to generate a proteinDB per\n",
      "                                group as specified in the filter_column,\n",
      "                                default is False\n",
      "  -h, --help                    Show this message and exit.\n"
     ]
    }
   ],
   "source": [
    "!pypgatk_cli.py cosmic-to-proteindb -h\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "edc86574",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "### Translate cosmic mutations\n",
    "To translate all mutations, we just need to specify paths for the mutations file and the transcripts file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "62fc4e0e",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n",
      "/Users/husen/opt/anaconda3/envs/pygatk/lib/python3.8/site-packages/Bio/Seq.py:2606: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\"Partial codon, len(sequence) not a multiple of three. \"\n"
     ]
    }
   ],
   "source": [
    "\n",
    "!pypgatk_cli.py cosmic-to-proteindb -c config/cosmic_config.yaml -in cosmic_files/CosmicMutantExport.tsv -fa cosmic_files/All_COSMIC_Genes.fasta -out cosmic_proteinDB.fa\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ca0b043",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "source": [
    "Since the actual `CosmicMutantExport.tsv` file contains around 50 Million mutations, we needs to filter them according to our needs otherwise the resulting protein database becomes too large. \n",
    "\n",
    "As shown in the help message, the `cosmic-to-proteindb` provides a `--filter_column` parameter that can be used to select any column for filtering and then with the `--accepted_values` parameter we can list the desired values. \n",
    "\n",
    "Inspect the following features to identify a column that you want to use as a filter:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "469c9aec",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Columns that can be used to filter mutations: Gene name\n",
      "Accession Number\n",
      "Gene CDS length\n",
      "HGNC ID\n",
      "Sample name\n",
      "ID_sample\n",
      "ID_tumour\n",
      "Primary site\n",
      "Site subtype 1\n",
      "Site subtype 2\n",
      "Site subtype 3\n",
      "Primary histology\n",
      "Histology subtype 1\n",
      "Histology subtype 2\n",
      "Histology subtype 3\n",
      "Genome-wide screen\n",
      "GENOMIC_MUTATION_ID\n",
      "LEGACY_MUTATION_ID\n",
      "MUTATION_ID\n",
      "Mutation CDS\n",
      "Mutation AA\n",
      "Mutation Description\n",
      "Mutation zygosity\n",
      "LOH\n",
      "GRCh\n",
      "Mutation genome position\n",
      "Mutation strand\n",
      "Resistance Mutation\n",
      "Mutation somatic status\n",
      "Pubmed_PMID\n",
      "ID_STUDY\n",
      "Sample Type\n",
      "Tumour origin\n",
      "Age\n",
      "HGVSP\n",
      "HGVSC\n",
      "HGVSG\n"
     ]
    }
   ],
   "source": [
    "fields = '\\n'.join(open('cosmic_files/CosmicMutantExport.tsv').readline().strip().split('\\t'))\n",
    "print('Columns that can be used to filter mutations:', fields)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ba88be5",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "source": [
    "Lets use `Primary site` as the filter column and accept mutations from tumors in liver, lung or skin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "ba4b6fb6",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n",
      "/Users/husen/opt/anaconda3/envs/pygatk/lib/python3.8/site-packages/Bio/Seq.py:2606: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\"Partial codon, len(sequence) not a multiple of three. \"\n",
      "Generated Counter({'COSMIC': 5100}) variant proteins from 4899 variants\n"
     ]
    }
   ],
   "source": [
    "!pypgatk_cli.py cosmic-to-proteindb -c config/cosmic_config.yaml -in cosmic_files/CosmicMutantExport.tsv -fa cosmic_files/All_COSMIC_Genes.fasta -out cosmic_proteinDB_filtered.fa --filter_column 'Primary site' --accepted_values 'liver,lung,skin'\n",
    "\n",
    "print('Generated {} variant proteins from {} variants'.format(\n",
    "    Counter([str(x.id).split(':')[0] for x in SeqIO.parse('cosmic_proteinDB_filtered.fa', \"fasta\")]),\n",
    "    len(set([':'.join(str(x.id).split(':')[2::]) for x in SeqIO.parse('cosmic_proteinDB_filtered.fa', \"fasta\")]))\n",
    "    ))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c85a807",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "source": [
    "We could also generate mutated proteins of a particular gene, e.g. `CDH6`. If you open the mutations file with a text editor, the gene names can be seen in the first column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "a6787f5a",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n",
      "/Users/husen/opt/anaconda3/envs/pygatk/lib/python3.8/site-packages/Bio/Seq.py:2606: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\"Partial codon, len(sequence) not a multiple of three. \"\n",
      "Generated Counter({'COSMIC': 453}) variant proteins from 453 variants\n"
     ]
    }
   ],
   "source": [
    "!pypgatk_cli.py cosmic-to-proteindb -c config/cosmic_config.yaml -in cosmic_files/CosmicMutantExport.tsv -fa cosmic_files/All_COSMIC_Genes.fasta -out cosmic_proteinDB_filtered_gene.fa --filter_column 'Gene name' --accepted_values 'CDH6'\n",
    "\n",
    "print('Generated {} variant proteins from {} variants'.format(\n",
    "    Counter([str(x.id).split(':')[0] for x in SeqIO.parse('cosmic_proteinDB_filtered_gene.fa', \"fasta\")]),\n",
    "    len(set([':'.join(str(x.id).split(':')[2::]) for x in SeqIO.parse('cosmic_proteinDB_filtered_gene.fa', \"fasta\")]))\n",
    "    ))\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Raw Cell Format",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
